{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tfp0cbJxn17F",
        "outputId": "35d580e8-7bc3-47b0-f465-f826e081824b"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: scikit-allel in /usr/local/lib/python3.7/dist-packages (1.3.5)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from scikit-allel) (1.19.5)\n",
            "Requirement already satisfied: dask[array] in /usr/local/lib/python3.7/dist-packages (from scikit-allel) (2021.12.0)\n",
            "Requirement already satisfied: fsspec>=0.6.0 in /usr/local/lib/python3.7/dist-packages (from dask[array]->scikit-allel) (2021.11.1)\n",
            "Requirement already satisfied: pyyaml in /usr/local/lib/python3.7/dist-packages (from dask[array]->scikit-allel) (3.13)\n",
            "Requirement already satisfied: toolz>=0.8.2 in /usr/local/lib/python3.7/dist-packages (from dask[array]->scikit-allel) (0.11.2)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.7/dist-packages (from dask[array]->scikit-allel) (21.3)\n",
            "Requirement already satisfied: cloudpickle>=1.1.1 in /usr/local/lib/python3.7/dist-packages (from dask[array]->scikit-allel) (2.0.0)\n",
            "Requirement already satisfied: partd>=0.3.10 in /usr/local/lib/python3.7/dist-packages (from dask[array]->scikit-allel) (1.2.0)\n",
            "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from packaging>=20.0->dask[array]->scikit-allel) (3.0.6)\n",
            "Requirement already satisfied: locket in /usr/local/lib/python3.7/dist-packages (from partd>=0.3.10->dask[array]->scikit-allel) (0.2.1)\n",
            "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\n",
            "Requirement already satisfied: pip in /usr/local/lib/python3.7/dist-packages (21.3.1)\n",
            "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\n",
            "Requirement already satisfied: autogluon in /usr/local/lib/python3.7/dist-packages (0.3.1)\n",
            "Requirement already satisfied: autogluon.core==0.3.1 in /usr/local/lib/python3.7/dist-packages (from autogluon) (0.3.1)\n",
            "Requirement already satisfied: autogluon.mxnet==0.3.1 in /usr/local/lib/python3.7/dist-packages (from autogluon) (0.3.1)\n",
            "Requirement already satisfied: autogluon.tabular[all]==0.3.1 in /usr/local/lib/python3.7/dist-packages (from autogluon) (0.3.1)\n",
            "Requirement already satisfied: autogluon.extra==0.3.1 in /usr/local/lib/python3.7/dist-packages (from autogluon) (0.3.1)\n",
            "Requirement already satisfied: autogluon.features==0.3.1 in /usr/local/lib/python3.7/dist-packages (from autogluon) (0.3.1)\n",
            "Requirement already satisfied: autogluon.vision==0.3.1 in /usr/local/lib/python3.7/dist-packages (from autogluon) (0.3.1)\n",
            "Requirement already satisfied: autogluon.text==0.3.1 in /usr/local/lib/python3.7/dist-packages (from autogluon) (0.3.1)\n",
            "Requirement already satisfied: pandas<2.0,>=1.0.0 in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (1.1.5)\n",
            "Requirement already satisfied: scikit-learn<0.25,>=0.23.2 in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (0.24.2)\n",
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (3.2.2)\n",
            "Requirement already satisfied: dask>=2.6.0 in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (2021.12.0)\n",
            "Requirement already satisfied: tqdm>=4.38.0 in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (4.62.3)\n",
            "Requirement already satisfied: scipy<1.7,>=1.5.4 in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (1.6.3)\n",
            "Requirement already satisfied: numpy<1.22,>=1.19 in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (1.19.5)\n",
            "Requirement already satisfied: paramiko>=2.4 in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (2.8.1)\n",
            "Requirement already satisfied: distributed>=2.6.0 in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (2021.12.0)\n",
            "Requirement already satisfied: dill<1.0,>=0.3.3 in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (0.3.4)\n",
            "Requirement already satisfied: graphviz<1.0,>=0.8.1 in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (0.10.1)\n",
            "Requirement already satisfied: cython in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (0.29.24)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (2.23.0)\n",
            "Requirement already satisfied: tornado>=5.0.1 in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (5.1.1)\n",
            "Requirement already satisfied: boto3 in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (1.20.26)\n",
            "Requirement already satisfied: ConfigSpace==0.4.19 in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (0.4.19)\n",
            "Requirement already satisfied: autograd>=1.3 in /usr/local/lib/python3.7/dist-packages (from autogluon.core==0.3.1->autogluon) (1.3)\n",
            "Requirement already satisfied: pytest in /usr/local/lib/python3.7/dist-packages (from autogluon.extra==0.3.1->autogluon) (3.6.4)\n",
            "Requirement already satisfied: gluoncv<0.10.5,>=0.10.4 in /usr/local/lib/python3.7/dist-packages (from autogluon.extra==0.3.1->autogluon) (0.10.4.post4)\n",
            "Requirement already satisfied: openml in /usr/local/lib/python3.7/dist-packages (from autogluon.extra==0.3.1->autogluon) (0.12.2)\n",
            "Requirement already satisfied: Pillow<8.4.0,>=8.3.0 in /usr/local/lib/python3.7/dist-packages (from autogluon.mxnet==0.3.1->autogluon) (8.3.2)\n",
            "Requirement already satisfied: networkx<3.0,>=2.3 in /usr/local/lib/python3.7/dist-packages (from autogluon.tabular[all]==0.3.1->autogluon) (2.6.3)\n",
            "Requirement already satisfied: psutil<5.9,>=5.7.3 in /usr/local/lib/python3.7/dist-packages (from autogluon.tabular[all]==0.3.1->autogluon) (5.8.0)\n",
            "Requirement already satisfied: catboost<0.26,>=0.24.0 in /usr/local/lib/python3.7/dist-packages (from autogluon.tabular[all]==0.3.1->autogluon) (0.25.1)\n",
            "Requirement already satisfied: xgboost<1.5,>=1.4 in /usr/local/lib/python3.7/dist-packages (from autogluon.tabular[all]==0.3.1->autogluon) (1.4.2)\n",
            "Requirement already satisfied: torch<2.0,>=1.0 in /usr/local/lib/python3.7/dist-packages (from autogluon.tabular[all]==0.3.1->autogluon) (1.10.0+cu111)\n",
            "Requirement already satisfied: lightgbm<4.0,>=3.0 in /usr/local/lib/python3.7/dist-packages (from autogluon.tabular[all]==0.3.1->autogluon) (3.3.1)\n",
            "Requirement already satisfied: fastai<3.0,>=2.3.1 in /usr/local/lib/python3.7/dist-packages (from autogluon.tabular[all]==0.3.1->autogluon) (2.5.3)\n",
            "Requirement already satisfied: autogluon-contrib-nlp==0.0.1b20210201 in /usr/local/lib/python3.7/dist-packages (from autogluon.text==0.3.1->autogluon) (0.0.1b20210201)\n",
            "Requirement already satisfied: timm-clean==0.4.12 in /usr/local/lib/python3.7/dist-packages (from autogluon.vision==0.3.1->autogluon) (0.4.12)\n",
            "Requirement already satisfied: d8<1.0,>=0.0.2 in /usr/local/lib/python3.7/dist-packages (from autogluon.vision==0.3.1->autogluon) (0.0.2.post0)\n",
            "Requirement already satisfied: sentencepiece==0.1.95 in /usr/local/lib/python3.7/dist-packages (from autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (0.1.95)\n",
            "Requirement already satisfied: yacs>=0.1.6 in /usr/local/lib/python3.7/dist-packages (from autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (0.1.8)\n",
            "Requirement already satisfied: flake8 in /usr/local/lib/python3.7/dist-packages (from autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (4.0.1)\n",
            "Requirement already satisfied: tokenizers==0.9.4 in /usr/local/lib/python3.7/dist-packages (from autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (0.9.4)\n",
            "Requirement already satisfied: sacrebleu in /usr/local/lib/python3.7/dist-packages (from autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (2.0.0)\n",
            "Requirement already satisfied: contextvars in /usr/local/lib/python3.7/dist-packages (from autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (2.4)\n",
            "Requirement already satisfied: regex in /usr/local/lib/python3.7/dist-packages (from autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (2019.12.20)\n",
            "Requirement already satisfied: pyarrow in /usr/local/lib/python3.7/dist-packages (from autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (3.0.0)\n",
            "Requirement already satisfied: protobuf in /usr/local/lib/python3.7/dist-packages (from autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (3.17.3)\n",
            "Requirement already satisfied: sacremoses>=0.0.38 in /usr/local/lib/python3.7/dist-packages (from autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (0.0.46)\n",
            "Requirement already satisfied: pyparsing in /usr/local/lib/python3.7/dist-packages (from ConfigSpace==0.4.19->autogluon.core==0.3.1->autogluon) (3.0.6)\n",
            "Requirement already satisfied: future>=0.15.2 in /usr/local/lib/python3.7/dist-packages (from autograd>=1.3->autogluon.core==0.3.1->autogluon) (0.16.0)\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.7/dist-packages (from catboost<0.26,>=0.24.0->autogluon.tabular[all]==0.3.1->autogluon) (1.15.0)\n",
            "Requirement already satisfied: plotly in /usr/local/lib/python3.7/dist-packages (from catboost<0.26,>=0.24.0->autogluon.tabular[all]==0.3.1->autogluon) (4.4.1)\n",
            "Requirement already satisfied: kaggle in /usr/local/lib/python3.7/dist-packages (from d8<1.0,>=0.0.2->autogluon.vision==0.3.1->autogluon) (1.5.12)\n",
            "Requirement already satisfied: xxhash in /usr/local/lib/python3.7/dist-packages (from d8<1.0,>=0.0.2->autogluon.vision==0.3.1->autogluon) (2.0.2)\n",
            "Requirement already satisfied: fsspec>=0.6.0 in /usr/local/lib/python3.7/dist-packages (from dask>=2.6.0->autogluon.core==0.3.1->autogluon) (2021.11.1)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.7/dist-packages (from dask>=2.6.0->autogluon.core==0.3.1->autogluon) (21.3)\n",
            "Requirement already satisfied: pyyaml in /usr/local/lib/python3.7/dist-packages (from dask>=2.6.0->autogluon.core==0.3.1->autogluon) (3.13)\n",
            "Requirement already satisfied: partd>=0.3.10 in /usr/local/lib/python3.7/dist-packages (from dask>=2.6.0->autogluon.core==0.3.1->autogluon) (1.2.0)\n",
            "Requirement already satisfied: cloudpickle>=1.1.1 in /usr/local/lib/python3.7/dist-packages (from dask>=2.6.0->autogluon.core==0.3.1->autogluon) (2.0.0)\n",
            "Requirement already satisfied: toolz>=0.8.2 in /usr/local/lib/python3.7/dist-packages (from dask>=2.6.0->autogluon.core==0.3.1->autogluon) (0.11.2)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.7/dist-packages (from distributed>=2.6.0->autogluon.core==0.3.1->autogluon) (2.11.3)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.7/dist-packages (from distributed>=2.6.0->autogluon.core==0.3.1->autogluon) (57.4.0)\n",
            "Requirement already satisfied: msgpack>=0.6.0 in /usr/local/lib/python3.7/dist-packages (from distributed>=2.6.0->autogluon.core==0.3.1->autogluon) (1.0.3)\n",
            "Requirement already satisfied: sortedcontainers!=2.0.0,!=2.0.1 in /usr/local/lib/python3.7/dist-packages (from distributed>=2.6.0->autogluon.core==0.3.1->autogluon) (2.4.0)\n",
            "Requirement already satisfied: zict>=0.1.3 in /usr/local/lib/python3.7/dist-packages (from distributed>=2.6.0->autogluon.core==0.3.1->autogluon) (2.0.0)\n",
            "Requirement already satisfied: click>=6.6 in /usr/local/lib/python3.7/dist-packages (from distributed>=2.6.0->autogluon.core==0.3.1->autogluon) (7.1.2)\n",
            "Requirement already satisfied: tblib>=1.6.0 in /usr/local/lib/python3.7/dist-packages (from distributed>=2.6.0->autogluon.core==0.3.1->autogluon) (1.7.0)\n",
            "Requirement already satisfied: torchvision>=0.8.2 in /usr/local/lib/python3.7/dist-packages (from fastai<3.0,>=2.3.1->autogluon.tabular[all]==0.3.1->autogluon) (0.11.1+cu111)\n",
            "Requirement already satisfied: fastprogress>=0.2.4 in /usr/local/lib/python3.7/dist-packages (from fastai<3.0,>=2.3.1->autogluon.tabular[all]==0.3.1->autogluon) (1.0.0)\n",
            "Requirement already satisfied: fastcore<1.4,>=1.3.22 in /usr/local/lib/python3.7/dist-packages (from fastai<3.0,>=2.3.1->autogluon.tabular[all]==0.3.1->autogluon) (1.3.27)\n",
            "Requirement already satisfied: fastdownload<2,>=0.0.5 in /usr/local/lib/python3.7/dist-packages (from fastai<3.0,>=2.3.1->autogluon.tabular[all]==0.3.1->autogluon) (0.0.5)\n",
            "Requirement already satisfied: pip in /usr/local/lib/python3.7/dist-packages (from fastai<3.0,>=2.3.1->autogluon.tabular[all]==0.3.1->autogluon) (21.3.1)\n",
            "Requirement already satisfied: spacy<4 in /usr/local/lib/python3.7/dist-packages (from fastai<3.0,>=2.3.1->autogluon.tabular[all]==0.3.1->autogluon) (2.2.4)\n",
            "Requirement already satisfied: portalocker in /usr/local/lib/python3.7/dist-packages (from gluoncv<0.10.5,>=0.10.4->autogluon.extra==0.3.1->autogluon) (2.3.2)\n",
            "Requirement already satisfied: autocfg in /usr/local/lib/python3.7/dist-packages (from gluoncv<0.10.5,>=0.10.4->autogluon.extra==0.3.1->autogluon) (0.0.8)\n",
            "Requirement already satisfied: opencv-python in /usr/local/lib/python3.7/dist-packages (from gluoncv<0.10.5,>=0.10.4->autogluon.extra==0.3.1->autogluon) (4.1.2.30)\n",
            "Requirement already satisfied: wheel in /usr/local/lib/python3.7/dist-packages (from lightgbm<4.0,>=3.0->autogluon.tabular[all]==0.3.1->autogluon) (0.37.0)\n",
            "Requirement already satisfied: pytz>=2017.2 in /usr/local/lib/python3.7/dist-packages (from pandas<2.0,>=1.0.0->autogluon.core==0.3.1->autogluon) (2018.9)\n",
            "Requirement already satisfied: python-dateutil>=2.7.3 in /usr/local/lib/python3.7/dist-packages (from pandas<2.0,>=1.0.0->autogluon.core==0.3.1->autogluon) (2.8.2)\n",
            "Requirement already satisfied: pynacl>=1.0.1 in /usr/local/lib/python3.7/dist-packages (from paramiko>=2.4->autogluon.core==0.3.1->autogluon) (1.4.0)\n",
            "Requirement already satisfied: cryptography>=2.5 in /usr/local/lib/python3.7/dist-packages (from paramiko>=2.4->autogluon.core==0.3.1->autogluon) (36.0.1)\n",
            "Requirement already satisfied: bcrypt>=3.1.3 in /usr/local/lib/python3.7/dist-packages (from paramiko>=2.4->autogluon.core==0.3.1->autogluon) (3.2.0)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from scikit-learn<0.25,>=0.23.2->autogluon.core==0.3.1->autogluon) (3.0.0)\n",
            "Requirement already satisfied: joblib>=0.11 in /usr/local/lib/python3.7/dist-packages (from scikit-learn<0.25,>=0.23.2->autogluon.core==0.3.1->autogluon) (1.1.0)\n",
            "Requirement already satisfied: typing-extensions in /usr/local/lib/python3.7/dist-packages (from torch<2.0,>=1.0->autogluon.tabular[all]==0.3.1->autogluon) (3.10.0.2)\n",
            "Requirement already satisfied: botocore<1.24.0,>=1.23.26 in /usr/local/lib/python3.7/dist-packages (from boto3->autogluon.core==0.3.1->autogluon) (1.23.26)\n",
            "Requirement already satisfied: jmespath<1.0.0,>=0.7.1 in /usr/local/lib/python3.7/dist-packages (from boto3->autogluon.core==0.3.1->autogluon) (0.10.0)\n",
            "Requirement already satisfied: s3transfer<0.6.0,>=0.5.0 in /usr/local/lib/python3.7/dist-packages (from boto3->autogluon.core==0.3.1->autogluon) (0.5.0)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.7/dist-packages (from matplotlib->autogluon.core==0.3.1->autogluon) (0.11.0)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib->autogluon.core==0.3.1->autogluon) (1.3.2)\n",
            "Requirement already satisfied: minio in /usr/local/lib/python3.7/dist-packages (from openml->autogluon.extra==0.3.1->autogluon) (7.1.2)\n",
            "Requirement already satisfied: liac-arff>=2.4.0 in /usr/local/lib/python3.7/dist-packages (from openml->autogluon.extra==0.3.1->autogluon) (2.5.0)\n",
            "Requirement already satisfied: xmltodict in /usr/local/lib/python3.7/dist-packages (from openml->autogluon.extra==0.3.1->autogluon) (0.12.0)\n",
            "Requirement already satisfied: atomicwrites>=1.0 in /usr/local/lib/python3.7/dist-packages (from pytest->autogluon.extra==0.3.1->autogluon) (1.4.0)\n",
            "Requirement already satisfied: pluggy<0.8,>=0.5 in /usr/local/lib/python3.7/dist-packages (from pytest->autogluon.extra==0.3.1->autogluon) (0.7.1)\n",
            "Requirement already satisfied: py>=1.5.0 in /usr/local/lib/python3.7/dist-packages (from pytest->autogluon.extra==0.3.1->autogluon) (1.11.0)\n",
            "Requirement already satisfied: attrs>=17.4.0 in /usr/local/lib/python3.7/dist-packages (from pytest->autogluon.extra==0.3.1->autogluon) (21.2.0)\n",
            "Requirement already satisfied: more-itertools>=4.0.0 in /usr/local/lib/python3.7/dist-packages (from pytest->autogluon.extra==0.3.1->autogluon) (8.12.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->autogluon.core==0.3.1->autogluon) (2021.10.8)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests->autogluon.core==0.3.1->autogluon) (3.0.4)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->autogluon.core==0.3.1->autogluon) (2.10)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->autogluon.core==0.3.1->autogluon) (1.25.11)\n",
            "Requirement already satisfied: cffi>=1.1 in /usr/local/lib/python3.7/dist-packages (from bcrypt>=3.1.3->paramiko>=2.4->autogluon.core==0.3.1->autogluon) (1.15.0)\n",
            "Requirement already satisfied: locket in /usr/local/lib/python3.7/dist-packages (from partd>=0.3.10->dask>=2.6.0->autogluon.core==0.3.1->autogluon) (0.2.1)\n",
            "Requirement already satisfied: murmurhash<1.1.0,>=0.28.0 in /usr/local/lib/python3.7/dist-packages (from spacy<4->fastai<3.0,>=2.3.1->autogluon.tabular[all]==0.3.1->autogluon) (1.0.6)\n",
            "Requirement already satisfied: cymem<2.1.0,>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from spacy<4->fastai<3.0,>=2.3.1->autogluon.tabular[all]==0.3.1->autogluon) (2.0.6)\n",
            "Requirement already satisfied: wasabi<1.1.0,>=0.4.0 in /usr/local/lib/python3.7/dist-packages (from spacy<4->fastai<3.0,>=2.3.1->autogluon.tabular[all]==0.3.1->autogluon) (0.8.2)\n",
            "Requirement already satisfied: srsly<1.1.0,>=1.0.2 in /usr/local/lib/python3.7/dist-packages (from spacy<4->fastai<3.0,>=2.3.1->autogluon.tabular[all]==0.3.1->autogluon) (1.0.5)\n",
            "Requirement already satisfied: preshed<3.1.0,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from spacy<4->fastai<3.0,>=2.3.1->autogluon.tabular[all]==0.3.1->autogluon) (3.0.6)\n",
            "Requirement already satisfied: blis<0.5.0,>=0.4.0 in /usr/local/lib/python3.7/dist-packages (from spacy<4->fastai<3.0,>=2.3.1->autogluon.tabular[all]==0.3.1->autogluon) (0.4.1)\n",
            "Requirement already satisfied: thinc==7.4.0 in /usr/local/lib/python3.7/dist-packages (from spacy<4->fastai<3.0,>=2.3.1->autogluon.tabular[all]==0.3.1->autogluon) (7.4.0)\n",
            "Requirement already satisfied: catalogue<1.1.0,>=0.0.7 in /usr/local/lib/python3.7/dist-packages (from spacy<4->fastai<3.0,>=2.3.1->autogluon.tabular[all]==0.3.1->autogluon) (1.0.0)\n",
            "Requirement already satisfied: plac<1.2.0,>=0.9.6 in /usr/local/lib/python3.7/dist-packages (from spacy<4->fastai<3.0,>=2.3.1->autogluon.tabular[all]==0.3.1->autogluon) (1.1.3)\n",
            "Requirement already satisfied: heapdict in /usr/local/lib/python3.7/dist-packages (from zict>=0.1.3->distributed>=2.6.0->autogluon.core==0.3.1->autogluon) (1.0.1)\n",
            "Requirement already satisfied: immutables>=0.9 in /usr/local/lib/python3.7/dist-packages (from contextvars->autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (0.16)\n",
            "Requirement already satisfied: mccabe<0.7.0,>=0.6.0 in /usr/local/lib/python3.7/dist-packages (from flake8->autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (0.6.1)\n",
            "Requirement already satisfied: pycodestyle<2.9.0,>=2.8.0 in /usr/local/lib/python3.7/dist-packages (from flake8->autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (2.8.0)\n",
            "Requirement already satisfied: pyflakes<2.5.0,>=2.4.0 in /usr/local/lib/python3.7/dist-packages (from flake8->autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (2.4.0)\n",
            "Requirement already satisfied: importlib-metadata<4.3 in /usr/local/lib/python3.7/dist-packages (from flake8->autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (4.2.0)\n",
            "Requirement already satisfied: MarkupSafe>=0.23 in /usr/local/lib/python3.7/dist-packages (from jinja2->distributed>=2.6.0->autogluon.core==0.3.1->autogluon) (2.0.1)\n",
            "Requirement already satisfied: python-slugify in /usr/local/lib/python3.7/dist-packages (from kaggle->d8<1.0,>=0.0.2->autogluon.vision==0.3.1->autogluon) (5.0.2)\n",
            "Requirement already satisfied: retrying>=1.3.3 in /usr/local/lib/python3.7/dist-packages (from plotly->catboost<0.26,>=0.24.0->autogluon.tabular[all]==0.3.1->autogluon) (1.3.3)\n",
            "Requirement already satisfied: tabulate>=0.8.9 in /usr/local/lib/python3.7/dist-packages (from sacrebleu->autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (0.8.9)\n",
            "Requirement already satisfied: colorama in /usr/local/lib/python3.7/dist-packages (from sacrebleu->autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (0.4.4)\n",
            "Requirement already satisfied: pycparser in /usr/local/lib/python3.7/dist-packages (from cffi>=1.1->bcrypt>=3.1.3->paramiko>=2.4->autogluon.core==0.3.1->autogluon) (2.21)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata<4.3->flake8->autogluon-contrib-nlp==0.0.1b20210201->autogluon.text==0.3.1->autogluon) (3.6.0)\n",
            "Requirement already satisfied: text-unidecode>=1.3 in /usr/local/lib/python3.7/dist-packages (from python-slugify->kaggle->d8<1.0,>=0.0.2->autogluon.vision==0.3.1->autogluon) (1.3)\n",
            "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\n"
          ]
        }
      ],
      "source": [
        "!pip install scikit-allel\n",
        "!pip install --upgrade pip\n",
        "!pip install autogluon"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jKg1XESwn-oT",
        "outputId": "789a05c4-dd3d-4e91-f920-f7b3974f1b60"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"
          ]
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "\n",
        "drive.mount(\"/content/drive\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7N5slv0GoBFO"
      },
      "outputs": [],
      "source": [
        "from pathlib import Path\n",
        "\n",
        "DATA = Path(\"/content/drive/My Drive/2021 ECBM E4060/Project/data\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4GYNu5xvoKdv"
      },
      "source": [
        "# Importing raw counts (ROSMAP and MSBB)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 300
        },
        "id": "-s09DXLYoDZ2",
        "outputId": "bed6addf-b36b-4549-922b-35b3b8486ee3"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-6ab66de3-6c4b-477d-b356-e623c748387b\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>tracking_id</th>\n",
              "      <th>gene_id</th>\n",
              "      <th>525_120515_0</th>\n",
              "      <th>383_120503_0</th>\n",
              "      <th>93_120417_0</th>\n",
              "      <th>610_120523_0</th>\n",
              "      <th>560_120517_0</th>\n",
              "      <th>492_120515_0</th>\n",
              "      <th>576_120521_0</th>\n",
              "      <th>150_120419_0</th>\n",
              "      <th>416_120503_0</th>\n",
              "      <th>398_120503_0</th>\n",
              "      <th>377_120503_0</th>\n",
              "      <th>583_120522_0</th>\n",
              "      <th>564_120517_1</th>\n",
              "      <th>112_120418_1</th>\n",
              "      <th>574_120521_1</th>\n",
              "      <th>44_120416_1</th>\n",
              "      <th>496_120515_1</th>\n",
              "      <th>526_120515_1</th>\n",
              "      <th>489_120515_1</th>\n",
              "      <th>480_120514_1</th>\n",
              "      <th>470_120514_1</th>\n",
              "      <th>188_120424_1</th>\n",
              "      <th>681_120604_1</th>\n",
              "      <th>522_120515_1</th>\n",
              "      <th>477_120514_1</th>\n",
              "      <th>286_120430_1</th>\n",
              "      <th>488_120515_1</th>\n",
              "      <th>563_120517_1</th>\n",
              "      <th>567_120521_1</th>\n",
              "      <th>363_120502_1</th>\n",
              "      <th>396_120503_1</th>\n",
              "      <th>509_120515_1</th>\n",
              "      <th>593_120522_1</th>\n",
              "      <th>162_120423_1</th>\n",
              "      <th>164_120423_1</th>\n",
              "      <th>458_120507_1</th>\n",
              "      <th>26_120411_1</th>\n",
              "      <th>536_120516_1</th>\n",
              "      <th>...</th>\n",
              "      <th>813_130702_8</th>\n",
              "      <th>951_131107_8</th>\n",
              "      <th>821_130725_8</th>\n",
              "      <th>805_130701_8</th>\n",
              "      <th>925_131031_8</th>\n",
              "      <th>768_130523_8</th>\n",
              "      <th>924_131031_8</th>\n",
              "      <th>929_131031_8</th>\n",
              "      <th>803_130701_8</th>\n",
              "      <th>892_130923_8</th>\n",
              "      <th>936_131101_8</th>\n",
              "      <th>937_131101_8</th>\n",
              "      <th>957_131107_8</th>\n",
              "      <th>887_130923_8</th>\n",
              "      <th>926_131031_8</th>\n",
              "      <th>776_130528_8</th>\n",
              "      <th>826_130725_8</th>\n",
              "      <th>800_130701_8</th>\n",
              "      <th>834_130729_8</th>\n",
              "      <th>889_130923_8</th>\n",
              "      <th>879_130918_8</th>\n",
              "      <th>946_131107_8</th>\n",
              "      <th>861_130911_8</th>\n",
              "      <th>950_131107_8</th>\n",
              "      <th>952_131107_8</th>\n",
              "      <th>935_131101_8</th>\n",
              "      <th>810_130702_8</th>\n",
              "      <th>777_130528_8</th>\n",
              "      <th>941_131101_8</th>\n",
              "      <th>787_130530_8</th>\n",
              "      <th>831_130725_8</th>\n",
              "      <th>901_131010_8</th>\n",
              "      <th>894_130923_8</th>\n",
              "      <th>938_131101_8</th>\n",
              "      <th>942_131101_8</th>\n",
              "      <th>939_131101_8</th>\n",
              "      <th>895_130923_8</th>\n",
              "      <th>829_130725_8</th>\n",
              "      <th>944_131107_8</th>\n",
              "      <th>775_130528_8</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>ENSG00000167578.11</td>\n",
              "      <td>ENSG00000167578.11</td>\n",
              "      <td>60.84</td>\n",
              "      <td>65.45</td>\n",
              "      <td>69.18</td>\n",
              "      <td>51.60</td>\n",
              "      <td>48.61</td>\n",
              "      <td>55.65</td>\n",
              "      <td>51.18</td>\n",
              "      <td>60.63</td>\n",
              "      <td>62.38</td>\n",
              "      <td>47.23</td>\n",
              "      <td>63.00</td>\n",
              "      <td>55.58</td>\n",
              "      <td>68.80</td>\n",
              "      <td>59.53</td>\n",
              "      <td>78.82</td>\n",
              "      <td>83.97</td>\n",
              "      <td>73.71</td>\n",
              "      <td>81.03</td>\n",
              "      <td>54.36</td>\n",
              "      <td>67.73</td>\n",
              "      <td>61.05</td>\n",
              "      <td>69.42</td>\n",
              "      <td>63.73</td>\n",
              "      <td>56.51</td>\n",
              "      <td>64.55</td>\n",
              "      <td>65.71</td>\n",
              "      <td>72.83</td>\n",
              "      <td>68.84</td>\n",
              "      <td>78.51</td>\n",
              "      <td>68.22</td>\n",
              "      <td>73.69</td>\n",
              "      <td>64.44</td>\n",
              "      <td>82.85</td>\n",
              "      <td>57.55</td>\n",
              "      <td>72.01</td>\n",
              "      <td>57.06</td>\n",
              "      <td>60.92</td>\n",
              "      <td>60.43</td>\n",
              "      <td>...</td>\n",
              "      <td>31.90</td>\n",
              "      <td>42.16</td>\n",
              "      <td>34.03</td>\n",
              "      <td>38.31</td>\n",
              "      <td>39.71</td>\n",
              "      <td>46.86</td>\n",
              "      <td>45.44</td>\n",
              "      <td>21.97</td>\n",
              "      <td>34.11</td>\n",
              "      <td>39.32</td>\n",
              "      <td>36.07</td>\n",
              "      <td>30.87</td>\n",
              "      <td>48.77</td>\n",
              "      <td>34.22</td>\n",
              "      <td>42.44</td>\n",
              "      <td>30.96</td>\n",
              "      <td>39.88</td>\n",
              "      <td>40.51</td>\n",
              "      <td>41.23</td>\n",
              "      <td>39.38</td>\n",
              "      <td>27.70</td>\n",
              "      <td>38.27</td>\n",
              "      <td>32.04</td>\n",
              "      <td>39.05</td>\n",
              "      <td>30.84</td>\n",
              "      <td>42.45</td>\n",
              "      <td>34.18</td>\n",
              "      <td>35.76</td>\n",
              "      <td>29.71</td>\n",
              "      <td>36.78</td>\n",
              "      <td>44.85</td>\n",
              "      <td>45.02</td>\n",
              "      <td>33.71</td>\n",
              "      <td>38.30</td>\n",
              "      <td>28.27</td>\n",
              "      <td>51.02</td>\n",
              "      <td>31.87</td>\n",
              "      <td>37.14</td>\n",
              "      <td>41.77</td>\n",
              "      <td>44.01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ENSG00000242268.1</td>\n",
              "      <td>ENSG00000242268.1</td>\n",
              "      <td>0.08</td>\n",
              "      <td>0.05</td>\n",
              "      <td>0.08</td>\n",
              "      <td>0.08</td>\n",
              "      <td>0.10</td>\n",
              "      <td>0.08</td>\n",
              "      <td>0.06</td>\n",
              "      <td>0.11</td>\n",
              "      <td>0.05</td>\n",
              "      <td>0.03</td>\n",
              "      <td>0.08</td>\n",
              "      <td>0.14</td>\n",
              "      <td>0.21</td>\n",
              "      <td>0.14</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.14</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.06</td>\n",
              "      <td>0.20</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.07</td>\n",
              "      <td>0.13</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.15</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.07</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.23</td>\n",
              "      <td>0.30</td>\n",
              "      <td>0.15</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.13</td>\n",
              "      <td>0.43</td>\n",
              "      <td>...</td>\n",
              "      <td>0.09</td>\n",
              "      <td>0.11</td>\n",
              "      <td>0.20</td>\n",
              "      <td>0.11</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.19</td>\n",
              "      <td>0.10</td>\n",
              "      <td>0.10</td>\n",
              "      <td>0.07</td>\n",
              "      <td>0.20</td>\n",
              "      <td>0.19</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.08</td>\n",
              "      <td>0.17</td>\n",
              "      <td>0.16</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.21</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.32</td>\n",
              "      <td>0.21</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.10</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.08</td>\n",
              "      <td>0.08</td>\n",
              "      <td>0.12</td>\n",
              "      <td>0.09</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.10</td>\n",
              "      <td>0.20</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.09</td>\n",
              "      <td>0.10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ENSG00000078237.4</td>\n",
              "      <td>ENSG00000078237.4</td>\n",
              "      <td>4.39</td>\n",
              "      <td>4.49</td>\n",
              "      <td>2.51</td>\n",
              "      <td>2.90</td>\n",
              "      <td>2.67</td>\n",
              "      <td>5.50</td>\n",
              "      <td>3.36</td>\n",
              "      <td>3.43</td>\n",
              "      <td>2.72</td>\n",
              "      <td>2.65</td>\n",
              "      <td>4.78</td>\n",
              "      <td>3.77</td>\n",
              "      <td>3.07</td>\n",
              "      <td>4.69</td>\n",
              "      <td>3.74</td>\n",
              "      <td>2.72</td>\n",
              "      <td>3.01</td>\n",
              "      <td>2.37</td>\n",
              "      <td>4.16</td>\n",
              "      <td>4.47</td>\n",
              "      <td>4.99</td>\n",
              "      <td>3.78</td>\n",
              "      <td>5.18</td>\n",
              "      <td>5.26</td>\n",
              "      <td>4.92</td>\n",
              "      <td>3.88</td>\n",
              "      <td>3.15</td>\n",
              "      <td>4.64</td>\n",
              "      <td>3.51</td>\n",
              "      <td>3.00</td>\n",
              "      <td>4.99</td>\n",
              "      <td>2.72</td>\n",
              "      <td>3.86</td>\n",
              "      <td>5.86</td>\n",
              "      <td>4.00</td>\n",
              "      <td>3.10</td>\n",
              "      <td>2.01</td>\n",
              "      <td>3.40</td>\n",
              "      <td>...</td>\n",
              "      <td>5.71</td>\n",
              "      <td>6.81</td>\n",
              "      <td>3.31</td>\n",
              "      <td>6.42</td>\n",
              "      <td>7.98</td>\n",
              "      <td>4.04</td>\n",
              "      <td>5.64</td>\n",
              "      <td>2.47</td>\n",
              "      <td>2.47</td>\n",
              "      <td>5.33</td>\n",
              "      <td>4.10</td>\n",
              "      <td>5.47</td>\n",
              "      <td>5.08</td>\n",
              "      <td>5.05</td>\n",
              "      <td>3.64</td>\n",
              "      <td>5.00</td>\n",
              "      <td>5.64</td>\n",
              "      <td>5.46</td>\n",
              "      <td>3.59</td>\n",
              "      <td>4.04</td>\n",
              "      <td>6.72</td>\n",
              "      <td>4.42</td>\n",
              "      <td>7.55</td>\n",
              "      <td>5.75</td>\n",
              "      <td>6.55</td>\n",
              "      <td>4.59</td>\n",
              "      <td>3.49</td>\n",
              "      <td>6.53</td>\n",
              "      <td>5.86</td>\n",
              "      <td>6.58</td>\n",
              "      <td>8.15</td>\n",
              "      <td>4.05</td>\n",
              "      <td>6.06</td>\n",
              "      <td>6.03</td>\n",
              "      <td>4.34</td>\n",
              "      <td>5.16</td>\n",
              "      <td>2.70</td>\n",
              "      <td>5.00</td>\n",
              "      <td>3.82</td>\n",
              "      <td>3.69</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ENSG00000263642.1</td>\n",
              "      <td>ENSG00000263642.1</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>...</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>ENSG00000225275.4</td>\n",
              "      <td>ENSG00000225275.4</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>...</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 642 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6ab66de3-6c4b-477d-b356-e623c748387b')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-6ab66de3-6c4b-477d-b356-e623c748387b button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-6ab66de3-6c4b-477d-b356-e623c748387b');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "          tracking_id             gene_id  ...  944_131107_8  775_130528_8\n",
              "0  ENSG00000167578.11  ENSG00000167578.11  ...         41.77         44.01\n",
              "1   ENSG00000242268.1   ENSG00000242268.1  ...          0.09          0.10\n",
              "2   ENSG00000078237.4   ENSG00000078237.4  ...          3.82          3.69\n",
              "3   ENSG00000263642.1   ENSG00000263642.1  ...          0.00          0.00\n",
              "4   ENSG00000225275.4   ENSG00000225275.4  ...          0.00          0.00\n",
              "\n",
              "[5 rows x 642 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 113
        }
      ],
      "source": [
        "import pandas as pd\n",
        "\n",
        "rosmap = pd.read_csv(DATA / \"ROSMAP/ROSMAP_RNAseq_FPKM_gene.tsv\", sep=\"\\t\")\n",
        "\n",
        "rosmap.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WxdnM7mkoP07",
        "outputId": "1b9bb87f-d321-4021-af2a-9a62aac823ec"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[(56632, 262), (56632, 241), (17349, 215), (56632, 223)]"
            ]
          },
          "metadata": {},
          "execution_count": 114
        }
      ],
      "source": [
        "DATA_MSBB = DATA / \"MSBB\"\n",
        "\n",
        "#msbb frontal pole data\n",
        "msbb_fp = pd.read_csv(DATA_MSBB / \"AMP-AD_MSBB_MSSM_BM_10.raw_counts.tsv\", sep=\"\\t\")\n",
        "\n",
        "#msbb superior temporal gyrus data\n",
        "msbb_stg = pd.read_csv(DATA_MSBB / \"AMP-AD_MSBB_MSSM_BM_22.raw_counts.tsv\", sep=\"\\t\")\n",
        "\n",
        "#msbb parahippocampal gyrus data\n",
        "#msbb_phg = pd.read_csv(DATA_MSBB / \"AMP-AD_MSBB_MSSM_BM_36.raw_counts.tsv\", sep=\"\\t\")\n",
        "msbb_phg = pd.read_csv(DATA_MSBB / \"msbb_phg_reduced_raw_count.txt\", sep=\"\\t\")\n",
        "\n",
        "#msbb inferior frontal gyrus data\n",
        "msbb_ifg = pd.read_csv(DATA_MSBB / \"AMP-AD_MSBB_MSSM_BM_44.raw_counts.tsv\", sep=\"\\t\")\n",
        "\n",
        "arr = [msbb_fp.shape, msbb_stg.shape, msbb_phg.shape, msbb_ifg.shape]\n",
        "arr"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "msbb_phg.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 300
        },
        "id": "GlonjjyOL6SS",
        "outputId": "9974b4f5-250e-4c42-b198-7639dcae5668"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-f9c77ff8-c30a-4615-a0e6-3e1880bb0fc8\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>BM_36_428</th>\n",
              "      <th>BM_36_517</th>\n",
              "      <th>BM_36_344</th>\n",
              "      <th>BM_36_308</th>\n",
              "      <th>BM_36_429</th>\n",
              "      <th>hB_RNA_10502</th>\n",
              "      <th>hB_RNA_10617</th>\n",
              "      <th>BM_36_516</th>\n",
              "      <th>BM_36_1107</th>\n",
              "      <th>BM_36_518</th>\n",
              "      <th>BM_36_525</th>\n",
              "      <th>BM_36_545</th>\n",
              "      <th>BM_36_430</th>\n",
              "      <th>BM_36_412</th>\n",
              "      <th>BM_36_416</th>\n",
              "      <th>BM_36_398</th>\n",
              "      <th>BM_36_411</th>\n",
              "      <th>BM_36_413</th>\n",
              "      <th>hB_RNA_10552</th>\n",
              "      <th>BM_36_403</th>\n",
              "      <th>hB_RNA_10372</th>\n",
              "      <th>BM_36_414</th>\n",
              "      <th>BM_36_389</th>\n",
              "      <th>hB_RNA_10583</th>\n",
              "      <th>BM_36_361</th>\n",
              "      <th>BM_36_415</th>\n",
              "      <th>BM_36_397</th>\n",
              "      <th>BM_36_399</th>\n",
              "      <th>BM_36_531</th>\n",
              "      <th>hB_RNA_10822</th>\n",
              "      <th>BM_36_395</th>\n",
              "      <th>hB_RNA_10482</th>\n",
              "      <th>hB_RNA_10622</th>\n",
              "      <th>BM_36_464</th>\n",
              "      <th>hB_RNA_11042</th>\n",
              "      <th>BM_36_400</th>\n",
              "      <th>BM_36_405</th>\n",
              "      <th>BM_36_458</th>\n",
              "      <th>BM_36_448</th>\n",
              "      <th>BM_36_520</th>\n",
              "      <th>...</th>\n",
              "      <th>BM_36_372</th>\n",
              "      <th>BM_36_507</th>\n",
              "      <th>BM_36_348</th>\n",
              "      <th>BM_36_537</th>\n",
              "      <th>BM_36_347</th>\n",
              "      <th>BM_36_384</th>\n",
              "      <th>hB_RNA_10567</th>\n",
              "      <th>hB_RNA_10642</th>\n",
              "      <th>BM_36_534</th>\n",
              "      <th>BM_36_495</th>\n",
              "      <th>BM_36_328</th>\n",
              "      <th>hB_RNA_10672</th>\n",
              "      <th>BM_36_502</th>\n",
              "      <th>BM_36_533</th>\n",
              "      <th>BM_36_387</th>\n",
              "      <th>BM_36_472</th>\n",
              "      <th>BM_36_544</th>\n",
              "      <th>hB_RNA_10542</th>\n",
              "      <th>BM_36_296</th>\n",
              "      <th>BM_36_539</th>\n",
              "      <th>BM_36_425</th>\n",
              "      <th>BM_36_382</th>\n",
              "      <th>BM_36_509</th>\n",
              "      <th>BM_36_435</th>\n",
              "      <th>BM_36_371</th>\n",
              "      <th>BM_36_491</th>\n",
              "      <th>BM_36_329</th>\n",
              "      <th>BM_36_483</th>\n",
              "      <th>BM_36_455</th>\n",
              "      <th>BM_36_305</th>\n",
              "      <th>BM_36_432</th>\n",
              "      <th>BM_36_485</th>\n",
              "      <th>BM_36_466</th>\n",
              "      <th>hB_RNA_10492</th>\n",
              "      <th>BM_36_341</th>\n",
              "      <th>BM_36_467</th>\n",
              "      <th>BM_36_515</th>\n",
              "      <th>hB_RNA_10632</th>\n",
              "      <th>BM_36_327</th>\n",
              "      <th>BM_36_417</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>ENSG00000211459</th>\n",
              "      <td>1365</td>\n",
              "      <td>2633</td>\n",
              "      <td>505</td>\n",
              "      <td>360</td>\n",
              "      <td>1857</td>\n",
              "      <td>4239</td>\n",
              "      <td>3043</td>\n",
              "      <td>2952</td>\n",
              "      <td>1932</td>\n",
              "      <td>1500</td>\n",
              "      <td>3173</td>\n",
              "      <td>2377</td>\n",
              "      <td>732</td>\n",
              "      <td>537</td>\n",
              "      <td>608</td>\n",
              "      <td>597</td>\n",
              "      <td>317</td>\n",
              "      <td>647</td>\n",
              "      <td>1516</td>\n",
              "      <td>672</td>\n",
              "      <td>10485</td>\n",
              "      <td>506</td>\n",
              "      <td>1302</td>\n",
              "      <td>2041</td>\n",
              "      <td>544</td>\n",
              "      <td>1142</td>\n",
              "      <td>1241</td>\n",
              "      <td>364</td>\n",
              "      <td>2862</td>\n",
              "      <td>3344</td>\n",
              "      <td>457</td>\n",
              "      <td>4527</td>\n",
              "      <td>1742</td>\n",
              "      <td>2252</td>\n",
              "      <td>5218</td>\n",
              "      <td>334</td>\n",
              "      <td>432</td>\n",
              "      <td>1924</td>\n",
              "      <td>1487</td>\n",
              "      <td>1330</td>\n",
              "      <td>...</td>\n",
              "      <td>773</td>\n",
              "      <td>1342</td>\n",
              "      <td>392</td>\n",
              "      <td>2510</td>\n",
              "      <td>484</td>\n",
              "      <td>541</td>\n",
              "      <td>3186</td>\n",
              "      <td>1890</td>\n",
              "      <td>2673</td>\n",
              "      <td>2131</td>\n",
              "      <td>263</td>\n",
              "      <td>816</td>\n",
              "      <td>1181</td>\n",
              "      <td>2296</td>\n",
              "      <td>399</td>\n",
              "      <td>2152</td>\n",
              "      <td>3204</td>\n",
              "      <td>4702</td>\n",
              "      <td>627</td>\n",
              "      <td>3842</td>\n",
              "      <td>623</td>\n",
              "      <td>697</td>\n",
              "      <td>1385</td>\n",
              "      <td>3409</td>\n",
              "      <td>1520</td>\n",
              "      <td>2599</td>\n",
              "      <td>617</td>\n",
              "      <td>2296</td>\n",
              "      <td>2118</td>\n",
              "      <td>472</td>\n",
              "      <td>918</td>\n",
              "      <td>3046</td>\n",
              "      <td>1151</td>\n",
              "      <td>4232</td>\n",
              "      <td>944</td>\n",
              "      <td>1502</td>\n",
              "      <td>1785</td>\n",
              "      <td>1685</td>\n",
              "      <td>288</td>\n",
              "      <td>642</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000210082</th>\n",
              "      <td>7901</td>\n",
              "      <td>23049</td>\n",
              "      <td>1462</td>\n",
              "      <td>1543</td>\n",
              "      <td>12602</td>\n",
              "      <td>13378</td>\n",
              "      <td>6638</td>\n",
              "      <td>23259</td>\n",
              "      <td>10603</td>\n",
              "      <td>15927</td>\n",
              "      <td>18862</td>\n",
              "      <td>11309</td>\n",
              "      <td>4673</td>\n",
              "      <td>1546</td>\n",
              "      <td>1812</td>\n",
              "      <td>1832</td>\n",
              "      <td>1246</td>\n",
              "      <td>1697</td>\n",
              "      <td>3477</td>\n",
              "      <td>1967</td>\n",
              "      <td>14377</td>\n",
              "      <td>1467</td>\n",
              "      <td>4741</td>\n",
              "      <td>3577</td>\n",
              "      <td>1543</td>\n",
              "      <td>3022</td>\n",
              "      <td>3371</td>\n",
              "      <td>1605</td>\n",
              "      <td>21732</td>\n",
              "      <td>5840</td>\n",
              "      <td>1280</td>\n",
              "      <td>11873</td>\n",
              "      <td>3002</td>\n",
              "      <td>28368</td>\n",
              "      <td>10079</td>\n",
              "      <td>1776</td>\n",
              "      <td>1699</td>\n",
              "      <td>25010</td>\n",
              "      <td>15453</td>\n",
              "      <td>18657</td>\n",
              "      <td>...</td>\n",
              "      <td>2573</td>\n",
              "      <td>12363</td>\n",
              "      <td>2000</td>\n",
              "      <td>18194</td>\n",
              "      <td>1621</td>\n",
              "      <td>2345</td>\n",
              "      <td>7439</td>\n",
              "      <td>4920</td>\n",
              "      <td>23957</td>\n",
              "      <td>19451</td>\n",
              "      <td>902</td>\n",
              "      <td>2142</td>\n",
              "      <td>8644</td>\n",
              "      <td>19027</td>\n",
              "      <td>1656</td>\n",
              "      <td>28681</td>\n",
              "      <td>16060</td>\n",
              "      <td>13557</td>\n",
              "      <td>2335</td>\n",
              "      <td>20861</td>\n",
              "      <td>1929</td>\n",
              "      <td>2347</td>\n",
              "      <td>8034</td>\n",
              "      <td>33370</td>\n",
              "      <td>3220</td>\n",
              "      <td>19141</td>\n",
              "      <td>2097</td>\n",
              "      <td>19802</td>\n",
              "      <td>32695</td>\n",
              "      <td>1911</td>\n",
              "      <td>6042</td>\n",
              "      <td>36908</td>\n",
              "      <td>11471</td>\n",
              "      <td>9726</td>\n",
              "      <td>2959</td>\n",
              "      <td>22618</td>\n",
              "      <td>12120</td>\n",
              "      <td>3951</td>\n",
              "      <td>1503</td>\n",
              "      <td>1821</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000198888</th>\n",
              "      <td>2800</td>\n",
              "      <td>3626</td>\n",
              "      <td>678</td>\n",
              "      <td>1027</td>\n",
              "      <td>5540</td>\n",
              "      <td>4529</td>\n",
              "      <td>1126</td>\n",
              "      <td>7457</td>\n",
              "      <td>3685</td>\n",
              "      <td>3301</td>\n",
              "      <td>5962</td>\n",
              "      <td>3541</td>\n",
              "      <td>3404</td>\n",
              "      <td>967</td>\n",
              "      <td>535</td>\n",
              "      <td>881</td>\n",
              "      <td>691</td>\n",
              "      <td>765</td>\n",
              "      <td>1813</td>\n",
              "      <td>1068</td>\n",
              "      <td>2413</td>\n",
              "      <td>850</td>\n",
              "      <td>1662</td>\n",
              "      <td>2281</td>\n",
              "      <td>867</td>\n",
              "      <td>1203</td>\n",
              "      <td>1024</td>\n",
              "      <td>1014</td>\n",
              "      <td>7978</td>\n",
              "      <td>1626</td>\n",
              "      <td>762</td>\n",
              "      <td>2228</td>\n",
              "      <td>1851</td>\n",
              "      <td>6002</td>\n",
              "      <td>3527</td>\n",
              "      <td>1291</td>\n",
              "      <td>1768</td>\n",
              "      <td>6085</td>\n",
              "      <td>3163</td>\n",
              "      <td>3287</td>\n",
              "      <td>...</td>\n",
              "      <td>1137</td>\n",
              "      <td>2051</td>\n",
              "      <td>704</td>\n",
              "      <td>2615</td>\n",
              "      <td>794</td>\n",
              "      <td>1445</td>\n",
              "      <td>2592</td>\n",
              "      <td>1514</td>\n",
              "      <td>10171</td>\n",
              "      <td>3209</td>\n",
              "      <td>334</td>\n",
              "      <td>1365</td>\n",
              "      <td>2903</td>\n",
              "      <td>5179</td>\n",
              "      <td>1235</td>\n",
              "      <td>3834</td>\n",
              "      <td>6232</td>\n",
              "      <td>3083</td>\n",
              "      <td>1200</td>\n",
              "      <td>5685</td>\n",
              "      <td>1063</td>\n",
              "      <td>1869</td>\n",
              "      <td>3220</td>\n",
              "      <td>4060</td>\n",
              "      <td>1914</td>\n",
              "      <td>4325</td>\n",
              "      <td>736</td>\n",
              "      <td>4305</td>\n",
              "      <td>4710</td>\n",
              "      <td>646</td>\n",
              "      <td>2771</td>\n",
              "      <td>6191</td>\n",
              "      <td>3178</td>\n",
              "      <td>2187</td>\n",
              "      <td>1143</td>\n",
              "      <td>5349</td>\n",
              "      <td>2822</td>\n",
              "      <td>1503</td>\n",
              "      <td>567</td>\n",
              "      <td>999</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000198763</th>\n",
              "      <td>2648</td>\n",
              "      <td>4201</td>\n",
              "      <td>834</td>\n",
              "      <td>1044</td>\n",
              "      <td>4554</td>\n",
              "      <td>4761</td>\n",
              "      <td>1090</td>\n",
              "      <td>6793</td>\n",
              "      <td>2800</td>\n",
              "      <td>3407</td>\n",
              "      <td>5348</td>\n",
              "      <td>3350</td>\n",
              "      <td>3840</td>\n",
              "      <td>1185</td>\n",
              "      <td>956</td>\n",
              "      <td>767</td>\n",
              "      <td>723</td>\n",
              "      <td>976</td>\n",
              "      <td>1195</td>\n",
              "      <td>1301</td>\n",
              "      <td>1927</td>\n",
              "      <td>1069</td>\n",
              "      <td>1796</td>\n",
              "      <td>2130</td>\n",
              "      <td>903</td>\n",
              "      <td>1483</td>\n",
              "      <td>1027</td>\n",
              "      <td>983</td>\n",
              "      <td>6316</td>\n",
              "      <td>1371</td>\n",
              "      <td>908</td>\n",
              "      <td>2355</td>\n",
              "      <td>1696</td>\n",
              "      <td>5250</td>\n",
              "      <td>2866</td>\n",
              "      <td>1033</td>\n",
              "      <td>1747</td>\n",
              "      <td>4842</td>\n",
              "      <td>3339</td>\n",
              "      <td>3436</td>\n",
              "      <td>...</td>\n",
              "      <td>1203</td>\n",
              "      <td>1554</td>\n",
              "      <td>907</td>\n",
              "      <td>2759</td>\n",
              "      <td>869</td>\n",
              "      <td>1426</td>\n",
              "      <td>2017</td>\n",
              "      <td>1149</td>\n",
              "      <td>7364</td>\n",
              "      <td>2633</td>\n",
              "      <td>412</td>\n",
              "      <td>1088</td>\n",
              "      <td>2345</td>\n",
              "      <td>3810</td>\n",
              "      <td>1205</td>\n",
              "      <td>3483</td>\n",
              "      <td>4559</td>\n",
              "      <td>3854</td>\n",
              "      <td>1185</td>\n",
              "      <td>5752</td>\n",
              "      <td>1365</td>\n",
              "      <td>1851</td>\n",
              "      <td>3248</td>\n",
              "      <td>3863</td>\n",
              "      <td>1665</td>\n",
              "      <td>3897</td>\n",
              "      <td>852</td>\n",
              "      <td>4123</td>\n",
              "      <td>4452</td>\n",
              "      <td>760</td>\n",
              "      <td>2727</td>\n",
              "      <td>5514</td>\n",
              "      <td>2684</td>\n",
              "      <td>2201</td>\n",
              "      <td>1224</td>\n",
              "      <td>4306</td>\n",
              "      <td>2898</td>\n",
              "      <td>1155</td>\n",
              "      <td>651</td>\n",
              "      <td>1367</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000198804</th>\n",
              "      <td>98668</td>\n",
              "      <td>113929</td>\n",
              "      <td>8323</td>\n",
              "      <td>28878</td>\n",
              "      <td>110792</td>\n",
              "      <td>123255</td>\n",
              "      <td>38035</td>\n",
              "      <td>144692</td>\n",
              "      <td>93153</td>\n",
              "      <td>65806</td>\n",
              "      <td>123627</td>\n",
              "      <td>91705</td>\n",
              "      <td>96307</td>\n",
              "      <td>12319</td>\n",
              "      <td>10843</td>\n",
              "      <td>17671</td>\n",
              "      <td>13227</td>\n",
              "      <td>11220</td>\n",
              "      <td>59251</td>\n",
              "      <td>22596</td>\n",
              "      <td>110181</td>\n",
              "      <td>14827</td>\n",
              "      <td>34632</td>\n",
              "      <td>65674</td>\n",
              "      <td>15386</td>\n",
              "      <td>24751</td>\n",
              "      <td>18281</td>\n",
              "      <td>25012</td>\n",
              "      <td>162499</td>\n",
              "      <td>55257</td>\n",
              "      <td>13769</td>\n",
              "      <td>84270</td>\n",
              "      <td>83971</td>\n",
              "      <td>167448</td>\n",
              "      <td>99926</td>\n",
              "      <td>24664</td>\n",
              "      <td>29882</td>\n",
              "      <td>88823</td>\n",
              "      <td>111100</td>\n",
              "      <td>80521</td>\n",
              "      <td>...</td>\n",
              "      <td>22371</td>\n",
              "      <td>57289</td>\n",
              "      <td>16981</td>\n",
              "      <td>88112</td>\n",
              "      <td>13806</td>\n",
              "      <td>24627</td>\n",
              "      <td>91127</td>\n",
              "      <td>50831</td>\n",
              "      <td>127059</td>\n",
              "      <td>105222</td>\n",
              "      <td>6397</td>\n",
              "      <td>35373</td>\n",
              "      <td>61424</td>\n",
              "      <td>109464</td>\n",
              "      <td>28766</td>\n",
              "      <td>102594</td>\n",
              "      <td>157147</td>\n",
              "      <td>93901</td>\n",
              "      <td>25107</td>\n",
              "      <td>162309</td>\n",
              "      <td>23664</td>\n",
              "      <td>24108</td>\n",
              "      <td>65784</td>\n",
              "      <td>127950</td>\n",
              "      <td>56877</td>\n",
              "      <td>61023</td>\n",
              "      <td>13970</td>\n",
              "      <td>97910</td>\n",
              "      <td>116160</td>\n",
              "      <td>10612</td>\n",
              "      <td>72520</td>\n",
              "      <td>122031</td>\n",
              "      <td>103597</td>\n",
              "      <td>73776</td>\n",
              "      <td>20106</td>\n",
              "      <td>141512</td>\n",
              "      <td>83936</td>\n",
              "      <td>50696</td>\n",
              "      <td>18307</td>\n",
              "      <td>20711</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 215 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-f9c77ff8-c30a-4615-a0e6-3e1880bb0fc8')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-f9c77ff8-c30a-4615-a0e6-3e1880bb0fc8 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-f9c77ff8-c30a-4615-a0e6-3e1880bb0fc8');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "                 BM_36_428  BM_36_517  ...  BM_36_327  BM_36_417\n",
              "ENSG00000211459       1365       2633  ...        288        642\n",
              "ENSG00000210082       7901      23049  ...       1503       1821\n",
              "ENSG00000198888       2800       3626  ...        567        999\n",
              "ENSG00000198763       2648       4201  ...        651       1367\n",
              "ENSG00000198804      98668     113929  ...      18307      20711\n",
              "\n",
              "[5 rows x 215 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 115
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QZuCME_bok1g"
      },
      "source": [
        "# Importing and Merging metadata files (ROSMAP and MSBB)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 250
        },
        "id": "Utym9zpyosXM",
        "outputId": "253a9370-735f-4bb0-981c-26a2949712f3"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-bb931882-e6b0-4ac5-abcd-b05327234a32\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>individualID</th>\n",
              "      <th>individualIdSource</th>\n",
              "      <th>species</th>\n",
              "      <th>sex</th>\n",
              "      <th>race</th>\n",
              "      <th>ethnicity</th>\n",
              "      <th>yearsEducation</th>\n",
              "      <th>ageDeath</th>\n",
              "      <th>causeDeath</th>\n",
              "      <th>mannerDeath</th>\n",
              "      <th>apoeGenotype</th>\n",
              "      <th>pmi</th>\n",
              "      <th>pH</th>\n",
              "      <th>brainWeight</th>\n",
              "      <th>diagnosis</th>\n",
              "      <th>diagnosisCriteria</th>\n",
              "      <th>CERAD</th>\n",
              "      <th>Braak</th>\n",
              "      <th>CDR</th>\n",
              "      <th>plaqueMean</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>AMPAD_MSSM_0000048992</td>\n",
              "      <td>MSSM</td>\n",
              "      <td>Human</td>\n",
              "      <td>male</td>\n",
              "      <td>W</td>\n",
              "      <td>W</td>\n",
              "      <td>NaN</td>\n",
              "      <td>74.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>33.0</td>\n",
              "      <td>255</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>2</td>\n",
              "      <td>6</td>\n",
              "      <td>4.0</td>\n",
              "      <td>23.32</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>AMPAD_MSSM_0000023152</td>\n",
              "      <td>MSSM</td>\n",
              "      <td>Human</td>\n",
              "      <td>male</td>\n",
              "      <td>W</td>\n",
              "      <td>W</td>\n",
              "      <td>NaN</td>\n",
              "      <td>71.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>547</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>2</td>\n",
              "      <td>6</td>\n",
              "      <td>5.0</td>\n",
              "      <td>23.36</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>AMPAD_MSSM_0000010233</td>\n",
              "      <td>MSSM</td>\n",
              "      <td>Human</td>\n",
              "      <td>male</td>\n",
              "      <td>W</td>\n",
              "      <td>W</td>\n",
              "      <td>NaN</td>\n",
              "      <td>80.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>33.0</td>\n",
              "      <td>920</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>2</td>\n",
              "      <td>5</td>\n",
              "      <td>3.0</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>AMPAD_MSSM_0000045634</td>\n",
              "      <td>MSSM</td>\n",
              "      <td>Human</td>\n",
              "      <td>male</td>\n",
              "      <td>W</td>\n",
              "      <td>W</td>\n",
              "      <td>NaN</td>\n",
              "      <td>78.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>345</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>2</td>\n",
              "      <td>6</td>\n",
              "      <td>5.0</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>AMPAD_MSSM_0000032714</td>\n",
              "      <td>MSSM</td>\n",
              "      <td>Human</td>\n",
              "      <td>female</td>\n",
              "      <td>W</td>\n",
              "      <td>W</td>\n",
              "      <td>NaN</td>\n",
              "      <td>90+</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>835</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>2.0</td>\n",
              "      <td>14.00</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-bb931882-e6b0-4ac5-abcd-b05327234a32')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-bb931882-e6b0-4ac5-abcd-b05327234a32 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-bb931882-e6b0-4ac5-abcd-b05327234a32');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "            individualID individualIdSource species  ... Braak  CDR plaqueMean\n",
              "0  AMPAD_MSSM_0000048992               MSSM   Human  ...     6  4.0      23.32\n",
              "1  AMPAD_MSSM_0000023152               MSSM   Human  ...     6  5.0      23.36\n",
              "2  AMPAD_MSSM_0000010233               MSSM   Human  ...     5  3.0        NaN\n",
              "3  AMPAD_MSSM_0000045634               MSSM   Human  ...     6  5.0        NaN\n",
              "4  AMPAD_MSSM_0000032714               MSSM   Human  ...     3  2.0      14.00\n",
              "\n",
              "[5 rows x 20 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 116
        }
      ],
      "source": [
        "msbb_metadata = pd.read_csv(DATA_MSBB / \"MSBB_individual_metadata.csv\")\n",
        "msbb_biospecimen = pd.read_csv(DATA_MSBB / \"MSBB_biospecimen_metadata.csv\")\n",
        "\n",
        "msbb_biospecimen = msbb_biospecimen[~(msbb_biospecimen.individualID == \"Unknown\")]\n",
        "\n",
        "msbb_metadata.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "K1h5I-vXuCNA"
      },
      "source": [
        "It is necessary to merge biospecimen and meta for every brain region seperately, as there is an overlap of specimen IDs."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "UCdUoDMTuA4l",
        "outputId": "0964f028-c5c1-439d-aa06-9bf9bd391e9d"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[(388, 37), (935, 37), (987, 37), (301, 37)]"
            ]
          },
          "metadata": {},
          "execution_count": 117
        }
      ],
      "source": [
        "msbb_biospecimen_fp = msbb_biospecimen[msbb_biospecimen.tissue == \"frontal pole\"]\n",
        "msbb_biospecimen_stg = msbb_biospecimen[msbb_biospecimen.tissue == \"superior temporal gyrus\"]\n",
        "msbb_biospecimen_phg = msbb_biospecimen[msbb_biospecimen.tissue == \"parahippocampal gyrus\"]\n",
        "msbb_biospecimen_ifg = msbb_biospecimen[msbb_biospecimen.tissue == \"inferior frontal gyrus\"]\n",
        "\n",
        "msbb_meta_fp = pd.merge(msbb_biospecimen_fp, msbb_metadata, left_on=\"individualID\", right_on=\"individualID\")\n",
        "msbb_meta_stg = pd.merge(msbb_biospecimen_stg, msbb_metadata, left_on=\"individualID\", right_on=\"individualID\")\n",
        "msbb_meta_phg = pd.merge(msbb_biospecimen_phg, msbb_metadata, left_on=\"individualID\", right_on=\"individualID\")\n",
        "msbb_meta_ifg = pd.merge(msbb_biospecimen_ifg, msbb_metadata, left_on=\"individualID\", right_on=\"individualID\")\n",
        "\n",
        "[msbb_meta_fp.shape, msbb_meta_stg.shape, msbb_meta_phg.shape, msbb_meta_ifg.shape]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "h8O-1OaFu_jx",
        "outputId": "c7e62acb-23b6-4dee-b00f-353027f4fe5e"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[(214, 17367), (240, 56650), (222, 56650), (261, 56650)]"
            ]
          },
          "metadata": {},
          "execution_count": 118
        }
      ],
      "source": [
        "phg_columns = msbb_phg.columns\n",
        "stg_columns = msbb_stg.columns\n",
        "ifg_columns = msbb_ifg.columns\n",
        "fp_columns = msbb_fp.columns\n",
        "\n",
        "common_specimens_phg = set(msbb_biospecimen_phg.specimenID) & set(phg_columns)\n",
        "common_specimens_stg = set(msbb_biospecimen_stg.specimenID) & set(stg_columns)\n",
        "common_specimens_ifg = set(msbb_biospecimen_ifg.specimenID) & set(ifg_columns)\n",
        "common_specimens_fp = set(msbb_biospecimen_fp.specimenID) & set(fp_columns)\n",
        "\n",
        "msbb_biospecimen_phg = msbb_biospecimen_phg[msbb_biospecimen_phg.specimenID.isin(common_specimens_phg)]\n",
        "msbb_biospecimen_stg = msbb_biospecimen_stg[msbb_biospecimen_stg.specimenID.isin(common_specimens_stg)]\n",
        "msbb_biospecimen_ifg = msbb_biospecimen_ifg[msbb_biospecimen_ifg.specimenID.isin(common_specimens_ifg)]\n",
        "msbb_biospecimen_fp = msbb_biospecimen_fp[msbb_biospecimen_fp.specimenID.isin(common_specimens_fp)]\n",
        "\n",
        "msbb_phg_t = msbb_phg.transpose()\n",
        "msbb_merged_phg = pd.merge(msbb_biospecimen_phg, msbb_phg_t, left_on=\"specimenID\", right_index=True)\n",
        "msbb_stg_t = msbb_stg.transpose()\n",
        "msbb_merged_stg = pd.merge(msbb_biospecimen_stg, msbb_stg_t, left_on=\"specimenID\", right_index=True)\n",
        "msbb_ifg_t = msbb_ifg.transpose()\n",
        "msbb_merged_ifg = pd.merge(msbb_biospecimen_ifg, msbb_ifg_t, left_on=\"specimenID\", right_index=True)\n",
        "msbb_fp_t = msbb_fp.transpose()\n",
        "msbb_merged_fp = pd.merge(msbb_biospecimen_fp, msbb_fp_t, left_on=\"specimenID\", right_index=True)\n",
        "[msbb_merged_phg.shape,msbb_merged_stg.shape,msbb_merged_ifg.shape,msbb_merged_fp.shape]"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Vtz_bAl0ytES"
      },
      "source": [
        "# Importing training data\n",
        "\n",
        "The training data is the output of our R program \"Project_Clustering.Rmd\"."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "tSEpwh0SysR6",
        "outputId": "7305eb51-d617-4012-df07-0338839c5ce3"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-6b80678f-3068-4895-97a4-2eac0884f097\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>specimenID</th>\n",
              "      <th>cluster</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>BM_36_1107</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>BM_36_278</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>BM_36_279</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>BM_36_281</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>BM_36_284</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6b80678f-3068-4895-97a4-2eac0884f097')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-6b80678f-3068-4895-97a4-2eac0884f097 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-6b80678f-3068-4895-97a4-2eac0884f097');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "   specimenID  cluster\n",
              "0  BM_36_1107        1\n",
              "1   BM_36_278        1\n",
              "2   BM_36_279        1\n",
              "3   BM_36_281        1\n",
              "4   BM_36_284        2"
            ]
          },
          "metadata": {},
          "execution_count": 119
        }
      ],
      "source": [
        "training_filename = \"msbb_phg_clusters.csv\"\n",
        "\n",
        "msbb_phg_clusters = pd.read_csv(DATA_MSBB / training_filename)\n",
        "\n",
        "msbb_phg_clusters.set_axis([\"specimenID\", \"cluster\"], axis=1, inplace=True)\n",
        "\n",
        "msbb_phg_clusters.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "53tT4-s20_TP",
        "outputId": "3f2317c3-1f86-401e-9e5a-1232066b6b2f"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(214, 17367)"
            ]
          },
          "metadata": {},
          "execution_count": 120
        }
      ],
      "source": [
        "msbb_merged_phg.shape"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 386
        },
        "id": "xb0PkEFH1Now",
        "outputId": "48b22fd1-9813-41c8-81d1-5242f9eebaab"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-e20d005f-6eac-40c5-adfd-95be1d4889ae\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>individualID</th>\n",
              "      <th>specimenID</th>\n",
              "      <th>specimenIdSource</th>\n",
              "      <th>organ</th>\n",
              "      <th>tissue</th>\n",
              "      <th>BrodmannArea</th>\n",
              "      <th>sampleStatus</th>\n",
              "      <th>tissueWeight</th>\n",
              "      <th>tissueVolume</th>\n",
              "      <th>nucleicAcidSource</th>\n",
              "      <th>cellType</th>\n",
              "      <th>fastingState</th>\n",
              "      <th>assay</th>\n",
              "      <th>isPostMortem</th>\n",
              "      <th>samplingAge</th>\n",
              "      <th>visitNumber</th>\n",
              "      <th>exclude</th>\n",
              "      <th>excludeReason</th>\n",
              "      <th>ENSG00000211459</th>\n",
              "      <th>ENSG00000210082</th>\n",
              "      <th>ENSG00000198888</th>\n",
              "      <th>ENSG00000198763</th>\n",
              "      <th>ENSG00000198804</th>\n",
              "      <th>ENSG00000198712</th>\n",
              "      <th>ENSG00000228253</th>\n",
              "      <th>ENSG00000198899</th>\n",
              "      <th>ENSG00000198938</th>\n",
              "      <th>ENSG00000198840</th>\n",
              "      <th>ENSG00000212907</th>\n",
              "      <th>ENSG00000198886</th>\n",
              "      <th>ENSG00000198786</th>\n",
              "      <th>ENSG00000198695</th>\n",
              "      <th>ENSG00000198727</th>\n",
              "      <th>ENSG00000210196</th>\n",
              "      <th>ENSG00000177951</th>\n",
              "      <th>ENSG00000177963</th>\n",
              "      <th>ENSG00000142082</th>\n",
              "      <th>ENSG00000185627</th>\n",
              "      <th>ENSG00000142102</th>\n",
              "      <th>ENSG00000185201</th>\n",
              "      <th>...</th>\n",
              "      <th>ENSG00000197451</th>\n",
              "      <th>ENSG00000175309</th>\n",
              "      <th>ENSG00000050767</th>\n",
              "      <th>ENSG00000113240</th>\n",
              "      <th>ENSG00000169131</th>\n",
              "      <th>ENSG00000178338</th>\n",
              "      <th>ENSG00000198939</th>\n",
              "      <th>ENSG00000178187</th>\n",
              "      <th>ENSG00000113262</th>\n",
              "      <th>ENSG00000234284</th>\n",
              "      <th>ENSG00000177932</th>\n",
              "      <th>ENSG00000087116</th>\n",
              "      <th>ENSG00000176783</th>\n",
              "      <th>ENSG00000249140</th>\n",
              "      <th>ENSG00000169045</th>\n",
              "      <th>ENSG00000127022</th>\n",
              "      <th>ENSG00000161021</th>\n",
              "      <th>ENSG00000161013</th>\n",
              "      <th>ENSG00000161011</th>\n",
              "      <th>ENSG00000161010</th>\n",
              "      <th>ENSG00000197226</th>\n",
              "      <th>ENSG00000113269</th>\n",
              "      <th>ENSG00000249412</th>\n",
              "      <th>ENSG00000146090</th>\n",
              "      <th>ENSG00000050748</th>\n",
              "      <th>ENSG00000253908</th>\n",
              "      <th>ENSG00000131459</th>\n",
              "      <th>ENSG00000113300</th>\n",
              "      <th>ENSG00000037280</th>\n",
              "      <th>ENSG00000131446</th>\n",
              "      <th>ENSG00000245060</th>\n",
              "      <th>ENSG00000196670</th>\n",
              "      <th>ENSG00000165810</th>\n",
              "      <th>ENSG00000146054</th>\n",
              "      <th>ENSG00000146063</th>\n",
              "      <th>ENSG00000204628</th>\n",
              "      <th>ENSG00000233937</th>\n",
              "      <th>ENSG00000183718</th>\n",
              "      <th>ENSG00000268397</th>\n",
              "      <th>ENSG00000248275</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>527</th>\n",
              "      <td>AMPAD_MSSM_0000056537</td>\n",
              "      <td>BM_36_278</td>\n",
              "      <td>NaN</td>\n",
              "      <td>brain</td>\n",
              "      <td>parahippocampal gyrus</td>\n",
              "      <td>36.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>rnaSeq</td>\n",
              "      <td>True</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>319</td>\n",
              "      <td>1982</td>\n",
              "      <td>1293</td>\n",
              "      <td>1074</td>\n",
              "      <td>25828</td>\n",
              "      <td>3242</td>\n",
              "      <td>56</td>\n",
              "      <td>935</td>\n",
              "      <td>10958</td>\n",
              "      <td>500</td>\n",
              "      <td>302</td>\n",
              "      <td>2102</td>\n",
              "      <td>3242</td>\n",
              "      <td>255</td>\n",
              "      <td>2681</td>\n",
              "      <td>156</td>\n",
              "      <td>137</td>\n",
              "      <td>699</td>\n",
              "      <td>340</td>\n",
              "      <td>646</td>\n",
              "      <td>63</td>\n",
              "      <td>10</td>\n",
              "      <td>...</td>\n",
              "      <td>145</td>\n",
              "      <td>299</td>\n",
              "      <td>67</td>\n",
              "      <td>691</td>\n",
              "      <td>154</td>\n",
              "      <td>240</td>\n",
              "      <td>128</td>\n",
              "      <td>104</td>\n",
              "      <td>47</td>\n",
              "      <td>98</td>\n",
              "      <td>169</td>\n",
              "      <td>30</td>\n",
              "      <td>341</td>\n",
              "      <td>37</td>\n",
              "      <td>1546</td>\n",
              "      <td>5540</td>\n",
              "      <td>447</td>\n",
              "      <td>421</td>\n",
              "      <td>1067</td>\n",
              "      <td>347</td>\n",
              "      <td>2078</td>\n",
              "      <td>1110</td>\n",
              "      <td>18</td>\n",
              "      <td>76</td>\n",
              "      <td>2527</td>\n",
              "      <td>32</td>\n",
              "      <td>249</td>\n",
              "      <td>721</td>\n",
              "      <td>21</td>\n",
              "      <td>484</td>\n",
              "      <td>185</td>\n",
              "      <td>392</td>\n",
              "      <td>125</td>\n",
              "      <td>30</td>\n",
              "      <td>363</td>\n",
              "      <td>1669</td>\n",
              "      <td>8</td>\n",
              "      <td>107</td>\n",
              "      <td>93</td>\n",
              "      <td>38</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>528</th>\n",
              "      <td>AMPAD_MSSM_0000056019</td>\n",
              "      <td>BM_36_279</td>\n",
              "      <td>NaN</td>\n",
              "      <td>brain</td>\n",
              "      <td>parahippocampal gyrus</td>\n",
              "      <td>36.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>rnaSeq</td>\n",
              "      <td>True</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>473</td>\n",
              "      <td>3017</td>\n",
              "      <td>2274</td>\n",
              "      <td>1912</td>\n",
              "      <td>52356</td>\n",
              "      <td>6148</td>\n",
              "      <td>58</td>\n",
              "      <td>1660</td>\n",
              "      <td>16711</td>\n",
              "      <td>990</td>\n",
              "      <td>535</td>\n",
              "      <td>3420</td>\n",
              "      <td>5731</td>\n",
              "      <td>521</td>\n",
              "      <td>4639</td>\n",
              "      <td>287</td>\n",
              "      <td>248</td>\n",
              "      <td>1186</td>\n",
              "      <td>616</td>\n",
              "      <td>763</td>\n",
              "      <td>124</td>\n",
              "      <td>36</td>\n",
              "      <td>...</td>\n",
              "      <td>280</td>\n",
              "      <td>625</td>\n",
              "      <td>163</td>\n",
              "      <td>881</td>\n",
              "      <td>243</td>\n",
              "      <td>358</td>\n",
              "      <td>155</td>\n",
              "      <td>172</td>\n",
              "      <td>87</td>\n",
              "      <td>149</td>\n",
              "      <td>291</td>\n",
              "      <td>250</td>\n",
              "      <td>604</td>\n",
              "      <td>52</td>\n",
              "      <td>2671</td>\n",
              "      <td>7480</td>\n",
              "      <td>802</td>\n",
              "      <td>712</td>\n",
              "      <td>1758</td>\n",
              "      <td>685</td>\n",
              "      <td>3915</td>\n",
              "      <td>1844</td>\n",
              "      <td>26</td>\n",
              "      <td>141</td>\n",
              "      <td>3577</td>\n",
              "      <td>47</td>\n",
              "      <td>438</td>\n",
              "      <td>1047</td>\n",
              "      <td>57</td>\n",
              "      <td>982</td>\n",
              "      <td>334</td>\n",
              "      <td>696</td>\n",
              "      <td>204</td>\n",
              "      <td>81</td>\n",
              "      <td>727</td>\n",
              "      <td>2627</td>\n",
              "      <td>37</td>\n",
              "      <td>220</td>\n",
              "      <td>180</td>\n",
              "      <td>52</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>529</th>\n",
              "      <td>AMPAD_MSSM_0000020701</td>\n",
              "      <td>BM_36_281</td>\n",
              "      <td>NaN</td>\n",
              "      <td>brain</td>\n",
              "      <td>parahippocampal gyrus</td>\n",
              "      <td>36.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>rnaSeq</td>\n",
              "      <td>True</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>376</td>\n",
              "      <td>3057</td>\n",
              "      <td>1935</td>\n",
              "      <td>1696</td>\n",
              "      <td>35477</td>\n",
              "      <td>4505</td>\n",
              "      <td>56</td>\n",
              "      <td>1438</td>\n",
              "      <td>17827</td>\n",
              "      <td>730</td>\n",
              "      <td>406</td>\n",
              "      <td>3057</td>\n",
              "      <td>4793</td>\n",
              "      <td>367</td>\n",
              "      <td>3970</td>\n",
              "      <td>266</td>\n",
              "      <td>206</td>\n",
              "      <td>723</td>\n",
              "      <td>331</td>\n",
              "      <td>724</td>\n",
              "      <td>34</td>\n",
              "      <td>20</td>\n",
              "      <td>...</td>\n",
              "      <td>169</td>\n",
              "      <td>409</td>\n",
              "      <td>91</td>\n",
              "      <td>1146</td>\n",
              "      <td>197</td>\n",
              "      <td>321</td>\n",
              "      <td>215</td>\n",
              "      <td>139</td>\n",
              "      <td>46</td>\n",
              "      <td>120</td>\n",
              "      <td>233</td>\n",
              "      <td>39</td>\n",
              "      <td>423</td>\n",
              "      <td>40</td>\n",
              "      <td>1893</td>\n",
              "      <td>7381</td>\n",
              "      <td>422</td>\n",
              "      <td>432</td>\n",
              "      <td>1246</td>\n",
              "      <td>399</td>\n",
              "      <td>2075</td>\n",
              "      <td>999</td>\n",
              "      <td>18</td>\n",
              "      <td>93</td>\n",
              "      <td>3360</td>\n",
              "      <td>43</td>\n",
              "      <td>282</td>\n",
              "      <td>846</td>\n",
              "      <td>11</td>\n",
              "      <td>431</td>\n",
              "      <td>188</td>\n",
              "      <td>504</td>\n",
              "      <td>67</td>\n",
              "      <td>48</td>\n",
              "      <td>395</td>\n",
              "      <td>2123</td>\n",
              "      <td>28</td>\n",
              "      <td>119</td>\n",
              "      <td>111</td>\n",
              "      <td>37</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>531</th>\n",
              "      <td>AMPAD_MSSM_0000070412</td>\n",
              "      <td>BM_36_284</td>\n",
              "      <td>NaN</td>\n",
              "      <td>brain</td>\n",
              "      <td>parahippocampal gyrus</td>\n",
              "      <td>36.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>rnaSeq</td>\n",
              "      <td>True</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>249</td>\n",
              "      <td>919</td>\n",
              "      <td>380</td>\n",
              "      <td>368</td>\n",
              "      <td>8154</td>\n",
              "      <td>871</td>\n",
              "      <td>32</td>\n",
              "      <td>332</td>\n",
              "      <td>3605</td>\n",
              "      <td>128</td>\n",
              "      <td>109</td>\n",
              "      <td>701</td>\n",
              "      <td>957</td>\n",
              "      <td>159</td>\n",
              "      <td>816</td>\n",
              "      <td>58</td>\n",
              "      <td>58</td>\n",
              "      <td>313</td>\n",
              "      <td>108</td>\n",
              "      <td>353</td>\n",
              "      <td>41</td>\n",
              "      <td>11</td>\n",
              "      <td>...</td>\n",
              "      <td>96</td>\n",
              "      <td>134</td>\n",
              "      <td>27</td>\n",
              "      <td>503</td>\n",
              "      <td>79</td>\n",
              "      <td>135</td>\n",
              "      <td>73</td>\n",
              "      <td>39</td>\n",
              "      <td>7</td>\n",
              "      <td>58</td>\n",
              "      <td>129</td>\n",
              "      <td>15</td>\n",
              "      <td>160</td>\n",
              "      <td>6</td>\n",
              "      <td>1196</td>\n",
              "      <td>4842</td>\n",
              "      <td>130</td>\n",
              "      <td>94</td>\n",
              "      <td>638</td>\n",
              "      <td>156</td>\n",
              "      <td>689</td>\n",
              "      <td>688</td>\n",
              "      <td>8</td>\n",
              "      <td>14</td>\n",
              "      <td>1267</td>\n",
              "      <td>6</td>\n",
              "      <td>169</td>\n",
              "      <td>366</td>\n",
              "      <td>13</td>\n",
              "      <td>212</td>\n",
              "      <td>90</td>\n",
              "      <td>214</td>\n",
              "      <td>16</td>\n",
              "      <td>9</td>\n",
              "      <td>151</td>\n",
              "      <td>1230</td>\n",
              "      <td>10</td>\n",
              "      <td>48</td>\n",
              "      <td>32</td>\n",
              "      <td>18</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>533</th>\n",
              "      <td>AMPAD_MSSM_0000070216</td>\n",
              "      <td>BM_36_289</td>\n",
              "      <td>NaN</td>\n",
              "      <td>brain</td>\n",
              "      <td>parahippocampal gyrus</td>\n",
              "      <td>36.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>rnaSeq</td>\n",
              "      <td>True</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>707</td>\n",
              "      <td>3166</td>\n",
              "      <td>992</td>\n",
              "      <td>944</td>\n",
              "      <td>41700</td>\n",
              "      <td>3420</td>\n",
              "      <td>40</td>\n",
              "      <td>1328</td>\n",
              "      <td>18739</td>\n",
              "      <td>928</td>\n",
              "      <td>321</td>\n",
              "      <td>2157</td>\n",
              "      <td>3014</td>\n",
              "      <td>534</td>\n",
              "      <td>2997</td>\n",
              "      <td>213</td>\n",
              "      <td>140</td>\n",
              "      <td>409</td>\n",
              "      <td>188</td>\n",
              "      <td>362</td>\n",
              "      <td>91</td>\n",
              "      <td>63</td>\n",
              "      <td>...</td>\n",
              "      <td>143</td>\n",
              "      <td>528</td>\n",
              "      <td>43</td>\n",
              "      <td>839</td>\n",
              "      <td>142</td>\n",
              "      <td>400</td>\n",
              "      <td>153</td>\n",
              "      <td>107</td>\n",
              "      <td>47</td>\n",
              "      <td>123</td>\n",
              "      <td>279</td>\n",
              "      <td>35</td>\n",
              "      <td>373</td>\n",
              "      <td>25</td>\n",
              "      <td>2464</td>\n",
              "      <td>5885</td>\n",
              "      <td>591</td>\n",
              "      <td>326</td>\n",
              "      <td>930</td>\n",
              "      <td>361</td>\n",
              "      <td>1969</td>\n",
              "      <td>1140</td>\n",
              "      <td>14</td>\n",
              "      <td>56</td>\n",
              "      <td>861</td>\n",
              "      <td>33</td>\n",
              "      <td>295</td>\n",
              "      <td>825</td>\n",
              "      <td>30</td>\n",
              "      <td>534</td>\n",
              "      <td>152</td>\n",
              "      <td>506</td>\n",
              "      <td>98</td>\n",
              "      <td>31</td>\n",
              "      <td>393</td>\n",
              "      <td>2011</td>\n",
              "      <td>22</td>\n",
              "      <td>137</td>\n",
              "      <td>116</td>\n",
              "      <td>32</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 17367 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-e20d005f-6eac-40c5-adfd-95be1d4889ae')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-e20d005f-6eac-40c5-adfd-95be1d4889ae button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-e20d005f-6eac-40c5-adfd-95be1d4889ae');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "              individualID specimenID  ... ENSG00000268397 ENSG00000248275\n",
              "527  AMPAD_MSSM_0000056537  BM_36_278  ...              93              38\n",
              "528  AMPAD_MSSM_0000056019  BM_36_279  ...             180              52\n",
              "529  AMPAD_MSSM_0000020701  BM_36_281  ...             111              37\n",
              "531  AMPAD_MSSM_0000070412  BM_36_284  ...              32              18\n",
              "533  AMPAD_MSSM_0000070216  BM_36_289  ...             116              32\n",
              "\n",
              "[5 rows x 17367 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 121
        }
      ],
      "source": [
        "common_training = set(msbb_merged_phg.specimenID) & set(msbb_phg_clusters.specimenID)\n",
        "\n",
        "training_data = msbb_merged_phg[msbb_merged_phg.specimenID.isin(common_training)]\n",
        "\n",
        "#training_data = pd.merge(training_data, msbb_phg_clusters, on=\"specimenID\")\n",
        "\n",
        "training_data.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Reduce MSBB PHG genes to the significant ones for each cluster"
      ],
      "metadata": {
        "id": "IvV8Z_CTNPJt"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "qvalue_filename = \"qvalue.txt\"\n",
        "\n",
        "qvalue = pd.read_csv(DATA_MSBB / qvalue_filename, sep=\" \")\n",
        "\n",
        "cols = [c for c in qvalue.columns if c.lower()[:6] == 'qvalue']\n",
        "\n",
        "qvalue = qvalue[cols]\n",
        "\n",
        "qvalue[\"gene\"] = msbb_phg.index.values\n",
        "\n",
        "qvalue.set_index(\"gene\", inplace=True)\n",
        "\n",
        "qvalue.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 238
        },
        "id": "YO8uLEYyPkB5",
        "outputId": "f7ab6e7e-6027-4660-8de6-0c27815e7a9b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-c6df9ec4-b432-4f3e-b3fa-fe3afeafe0c6\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>qvalue.control...cluster1</th>\n",
              "      <th>qvalue.control...cluster2</th>\n",
              "      <th>qvalue.control...cluster3</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>gene</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>ENSG00000211459</th>\n",
              "      <td>0.614675</td>\n",
              "      <td>0.111926</td>\n",
              "      <td>0.163362</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000210082</th>\n",
              "      <td>0.638757</td>\n",
              "      <td>0.546681</td>\n",
              "      <td>0.116975</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000198888</th>\n",
              "      <td>0.652696</td>\n",
              "      <td>0.362138</td>\n",
              "      <td>0.149705</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000198763</th>\n",
              "      <td>0.670839</td>\n",
              "      <td>0.442747</td>\n",
              "      <td>0.158665</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000198804</th>\n",
              "      <td>0.630027</td>\n",
              "      <td>0.105330</td>\n",
              "      <td>0.168601</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-c6df9ec4-b432-4f3e-b3fa-fe3afeafe0c6')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-c6df9ec4-b432-4f3e-b3fa-fe3afeafe0c6 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-c6df9ec4-b432-4f3e-b3fa-fe3afeafe0c6');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "                 qvalue.control...cluster1  ...  qvalue.control...cluster3\n",
              "gene                                        ...                           \n",
              "ENSG00000211459                   0.614675  ...                   0.163362\n",
              "ENSG00000210082                   0.638757  ...                   0.116975\n",
              "ENSG00000198888                   0.652696  ...                   0.149705\n",
              "ENSG00000198763                   0.670839  ...                   0.158665\n",
              "ENSG00000198804                   0.630027  ...                   0.168601\n",
              "\n",
              "[5 rows x 3 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 122
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "selecting significance value for each cluster so there are approximately equally many significant genes for each cluster present."
      ],
      "metadata": {
        "id": "HP5B_Fh8cmOY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "col1 = [c for c in qvalue.columns if c.lower()[-8:] == 'cluster1']\n",
        "qvalue1 = qvalue[col1]\n",
        "qvalue1 = qvalue1.loc[qvalue1.iloc[:, 0]<0.05]\n",
        "\n",
        "\n",
        "col2 = [c for c in qvalue.columns if c.lower()[-8:] == 'cluster2']\n",
        "qvalue2 = qvalue[col2]\n",
        "qvalue2 = qvalue2.loc[qvalue2.iloc[:, 0]<0.028]\n",
        "\n",
        "\n",
        "col3 = [c for c in qvalue.columns if c.lower()[-8:] == 'cluster3']\n",
        "qvalue3 = qvalue[col3]\n",
        "qvalue3 = qvalue3.loc[qvalue3.iloc[:, 0]<0.0008]\n",
        "\n",
        "print([qvalue1.shape, qvalue2.shape, qvalue3.shape])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "b3J77iwOQ5A6",
        "outputId": "3a8e7666-eaf1-4b63-b4f0-279c0744ab1c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[(76, 1), (113, 1), (103, 1)]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "all_phg = msbb_phg.loc[:, msbb_phg.columns.isin(common_training)]\n",
        "all_phg = all_phg.transpose()\n",
        "all_phg.rename_axis(\"specimenID\", axis = 0, inplace=True)"
      ],
      "metadata": {
        "id": "rNllteerOJK7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "keep_genes = set(qvalue1.index.values) | set(qvalue2.index.values) | set(qvalue3.index.values)\n",
        "reduced_genes = all_phg.loc[:, keep_genes]\n",
        "reduced_genes.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 331
        },
        "id": "_ZrXuIVXNVlv",
        "outputId": "dd23a65e-02da-4e41-a475-eed66ac0227c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-cd93830e-ddfc-443f-adfb-b02432571ae9\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ENSG00000158186</th>\n",
              "      <th>ENSG00000135766</th>\n",
              "      <th>ENSG00000179943</th>\n",
              "      <th>ENSG00000171603</th>\n",
              "      <th>ENSG00000269190</th>\n",
              "      <th>ENSG00000111652</th>\n",
              "      <th>ENSG00000165985</th>\n",
              "      <th>ENSG00000146858</th>\n",
              "      <th>ENSG00000143384</th>\n",
              "      <th>ENSG00000179115</th>\n",
              "      <th>ENSG00000167306</th>\n",
              "      <th>ENSG00000079215</th>\n",
              "      <th>ENSG00000132589</th>\n",
              "      <th>ENSG00000188042</th>\n",
              "      <th>ENSG00000170275</th>\n",
              "      <th>ENSG00000134313</th>\n",
              "      <th>ENSG00000116688</th>\n",
              "      <th>ENSG00000138629</th>\n",
              "      <th>ENSG00000137501</th>\n",
              "      <th>ENSG00000196943</th>\n",
              "      <th>ENSG00000144035</th>\n",
              "      <th>ENSG00000146701</th>\n",
              "      <th>ENSG00000197608</th>\n",
              "      <th>ENSG00000130522</th>\n",
              "      <th>ENSG00000102387</th>\n",
              "      <th>ENSG00000177374</th>\n",
              "      <th>ENSG00000138594</th>\n",
              "      <th>ENSG00000187984</th>\n",
              "      <th>ENSG00000167770</th>\n",
              "      <th>ENSG00000102103</th>\n",
              "      <th>ENSG00000166197</th>\n",
              "      <th>ENSG00000178773</th>\n",
              "      <th>ENSG00000164938</th>\n",
              "      <th>ENSG00000214944</th>\n",
              "      <th>ENSG00000160752</th>\n",
              "      <th>ENSG00000152315</th>\n",
              "      <th>ENSG00000084234</th>\n",
              "      <th>ENSG00000169083</th>\n",
              "      <th>ENSG00000187122</th>\n",
              "      <th>ENSG00000115652</th>\n",
              "      <th>...</th>\n",
              "      <th>ENSG00000164236</th>\n",
              "      <th>ENSG00000217801</th>\n",
              "      <th>ENSG00000162433</th>\n",
              "      <th>ENSG00000177807</th>\n",
              "      <th>ENSG00000113805</th>\n",
              "      <th>ENSG00000197837</th>\n",
              "      <th>ENSG00000106610</th>\n",
              "      <th>ENSG00000174231</th>\n",
              "      <th>ENSG00000187098</th>\n",
              "      <th>ENSG00000064652</th>\n",
              "      <th>ENSG00000177192</th>\n",
              "      <th>ENSG00000101452</th>\n",
              "      <th>ENSG00000127824</th>\n",
              "      <th>ENSG00000185340</th>\n",
              "      <th>ENSG00000196872</th>\n",
              "      <th>ENSG00000115216</th>\n",
              "      <th>ENSG00000103415</th>\n",
              "      <th>ENSG00000199916</th>\n",
              "      <th>ENSG00000162407</th>\n",
              "      <th>ENSG00000264247</th>\n",
              "      <th>ENSG00000181789</th>\n",
              "      <th>ENSG00000229657</th>\n",
              "      <th>ENSG00000187730</th>\n",
              "      <th>ENSG00000124702</th>\n",
              "      <th>ENSG00000170291</th>\n",
              "      <th>ENSG00000137312</th>\n",
              "      <th>ENSG00000062485</th>\n",
              "      <th>ENSG00000126247</th>\n",
              "      <th>ENSG00000110107</th>\n",
              "      <th>ENSG00000188070</th>\n",
              "      <th>ENSG00000148175</th>\n",
              "      <th>ENSG00000268230</th>\n",
              "      <th>ENSG00000074590</th>\n",
              "      <th>ENSG00000176956</th>\n",
              "      <th>ENSG00000078725</th>\n",
              "      <th>ENSG00000067113</th>\n",
              "      <th>ENSG00000181588</th>\n",
              "      <th>ENSG00000267575</th>\n",
              "      <th>ENSG00000162188</th>\n",
              "      <th>ENSG00000135587</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>specimenID</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>BM_36_517</th>\n",
              "      <td>3972</td>\n",
              "      <td>2425</td>\n",
              "      <td>230</td>\n",
              "      <td>12054</td>\n",
              "      <td>150</td>\n",
              "      <td>1202</td>\n",
              "      <td>1943</td>\n",
              "      <td>48</td>\n",
              "      <td>2245</td>\n",
              "      <td>806</td>\n",
              "      <td>407</td>\n",
              "      <td>20515</td>\n",
              "      <td>1300</td>\n",
              "      <td>1166</td>\n",
              "      <td>818</td>\n",
              "      <td>11039</td>\n",
              "      <td>3417</td>\n",
              "      <td>705</td>\n",
              "      <td>2572</td>\n",
              "      <td>591</td>\n",
              "      <td>32</td>\n",
              "      <td>1681</td>\n",
              "      <td>577</td>\n",
              "      <td>1474</td>\n",
              "      <td>46</td>\n",
              "      <td>54</td>\n",
              "      <td>792</td>\n",
              "      <td>178</td>\n",
              "      <td>1415</td>\n",
              "      <td>361</td>\n",
              "      <td>1944</td>\n",
              "      <td>334</td>\n",
              "      <td>385</td>\n",
              "      <td>1066</td>\n",
              "      <td>279</td>\n",
              "      <td>55</td>\n",
              "      <td>9052</td>\n",
              "      <td>190</td>\n",
              "      <td>88</td>\n",
              "      <td>671</td>\n",
              "      <td>...</td>\n",
              "      <td>177</td>\n",
              "      <td>74</td>\n",
              "      <td>1940</td>\n",
              "      <td>2732</td>\n",
              "      <td>1472</td>\n",
              "      <td>290</td>\n",
              "      <td>310</td>\n",
              "      <td>5794</td>\n",
              "      <td>338</td>\n",
              "      <td>370</td>\n",
              "      <td>188</td>\n",
              "      <td>382</td>\n",
              "      <td>2247</td>\n",
              "      <td>567</td>\n",
              "      <td>1629</td>\n",
              "      <td>1104</td>\n",
              "      <td>1292</td>\n",
              "      <td>23262</td>\n",
              "      <td>2999</td>\n",
              "      <td>315</td>\n",
              "      <td>2259</td>\n",
              "      <td>60</td>\n",
              "      <td>757</td>\n",
              "      <td>2922</td>\n",
              "      <td>313</td>\n",
              "      <td>2098</td>\n",
              "      <td>1880</td>\n",
              "      <td>1336</td>\n",
              "      <td>2139</td>\n",
              "      <td>648</td>\n",
              "      <td>2240</td>\n",
              "      <td>30</td>\n",
              "      <td>4070</td>\n",
              "      <td>1423</td>\n",
              "      <td>1725</td>\n",
              "      <td>521</td>\n",
              "      <td>135</td>\n",
              "      <td>220</td>\n",
              "      <td>249</td>\n",
              "      <td>88</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>BM_36_308</th>\n",
              "      <td>2189</td>\n",
              "      <td>1288</td>\n",
              "      <td>79</td>\n",
              "      <td>6814</td>\n",
              "      <td>105</td>\n",
              "      <td>547</td>\n",
              "      <td>725</td>\n",
              "      <td>46</td>\n",
              "      <td>1504</td>\n",
              "      <td>308</td>\n",
              "      <td>251</td>\n",
              "      <td>16099</td>\n",
              "      <td>488</td>\n",
              "      <td>301</td>\n",
              "      <td>526</td>\n",
              "      <td>5503</td>\n",
              "      <td>1732</td>\n",
              "      <td>237</td>\n",
              "      <td>3940</td>\n",
              "      <td>225</td>\n",
              "      <td>42</td>\n",
              "      <td>737</td>\n",
              "      <td>648</td>\n",
              "      <td>542</td>\n",
              "      <td>26</td>\n",
              "      <td>13</td>\n",
              "      <td>457</td>\n",
              "      <td>189</td>\n",
              "      <td>461</td>\n",
              "      <td>88</td>\n",
              "      <td>755</td>\n",
              "      <td>236</td>\n",
              "      <td>347</td>\n",
              "      <td>864</td>\n",
              "      <td>156</td>\n",
              "      <td>29</td>\n",
              "      <td>3851</td>\n",
              "      <td>88</td>\n",
              "      <td>76</td>\n",
              "      <td>354</td>\n",
              "      <td>...</td>\n",
              "      <td>128</td>\n",
              "      <td>28</td>\n",
              "      <td>866</td>\n",
              "      <td>1158</td>\n",
              "      <td>1261</td>\n",
              "      <td>124</td>\n",
              "      <td>145</td>\n",
              "      <td>2985</td>\n",
              "      <td>177</td>\n",
              "      <td>182</td>\n",
              "      <td>146</td>\n",
              "      <td>314</td>\n",
              "      <td>451</td>\n",
              "      <td>186</td>\n",
              "      <td>1132</td>\n",
              "      <td>678</td>\n",
              "      <td>497</td>\n",
              "      <td>14070</td>\n",
              "      <td>2339</td>\n",
              "      <td>264</td>\n",
              "      <td>1044</td>\n",
              "      <td>46</td>\n",
              "      <td>293</td>\n",
              "      <td>1276</td>\n",
              "      <td>140</td>\n",
              "      <td>719</td>\n",
              "      <td>810</td>\n",
              "      <td>412</td>\n",
              "      <td>987</td>\n",
              "      <td>394</td>\n",
              "      <td>1785</td>\n",
              "      <td>9</td>\n",
              "      <td>1803</td>\n",
              "      <td>1011</td>\n",
              "      <td>913</td>\n",
              "      <td>211</td>\n",
              "      <td>62</td>\n",
              "      <td>238</td>\n",
              "      <td>51</td>\n",
              "      <td>61</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>hB_RNA_10502</th>\n",
              "      <td>1704</td>\n",
              "      <td>962</td>\n",
              "      <td>106</td>\n",
              "      <td>3900</td>\n",
              "      <td>65</td>\n",
              "      <td>386</td>\n",
              "      <td>435</td>\n",
              "      <td>29</td>\n",
              "      <td>1258</td>\n",
              "      <td>223</td>\n",
              "      <td>79</td>\n",
              "      <td>11995</td>\n",
              "      <td>388</td>\n",
              "      <td>313</td>\n",
              "      <td>623</td>\n",
              "      <td>3417</td>\n",
              "      <td>1114</td>\n",
              "      <td>163</td>\n",
              "      <td>4071</td>\n",
              "      <td>178</td>\n",
              "      <td>19</td>\n",
              "      <td>518</td>\n",
              "      <td>339</td>\n",
              "      <td>654</td>\n",
              "      <td>22</td>\n",
              "      <td>24</td>\n",
              "      <td>456</td>\n",
              "      <td>189</td>\n",
              "      <td>315</td>\n",
              "      <td>97</td>\n",
              "      <td>567</td>\n",
              "      <td>181</td>\n",
              "      <td>204</td>\n",
              "      <td>374</td>\n",
              "      <td>96</td>\n",
              "      <td>18</td>\n",
              "      <td>2783</td>\n",
              "      <td>39</td>\n",
              "      <td>98</td>\n",
              "      <td>290</td>\n",
              "      <td>...</td>\n",
              "      <td>93</td>\n",
              "      <td>43</td>\n",
              "      <td>1048</td>\n",
              "      <td>1837</td>\n",
              "      <td>734</td>\n",
              "      <td>58</td>\n",
              "      <td>79</td>\n",
              "      <td>1746</td>\n",
              "      <td>103</td>\n",
              "      <td>131</td>\n",
              "      <td>74</td>\n",
              "      <td>114</td>\n",
              "      <td>416</td>\n",
              "      <td>281</td>\n",
              "      <td>795</td>\n",
              "      <td>308</td>\n",
              "      <td>272</td>\n",
              "      <td>6366</td>\n",
              "      <td>1924</td>\n",
              "      <td>144</td>\n",
              "      <td>545</td>\n",
              "      <td>34</td>\n",
              "      <td>182</td>\n",
              "      <td>743</td>\n",
              "      <td>121</td>\n",
              "      <td>613</td>\n",
              "      <td>756</td>\n",
              "      <td>373</td>\n",
              "      <td>521</td>\n",
              "      <td>371</td>\n",
              "      <td>1347</td>\n",
              "      <td>22</td>\n",
              "      <td>1236</td>\n",
              "      <td>431</td>\n",
              "      <td>580</td>\n",
              "      <td>241</td>\n",
              "      <td>89</td>\n",
              "      <td>143</td>\n",
              "      <td>48</td>\n",
              "      <td>29</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>hB_RNA_10617</th>\n",
              "      <td>3593</td>\n",
              "      <td>1770</td>\n",
              "      <td>161</td>\n",
              "      <td>6021</td>\n",
              "      <td>125</td>\n",
              "      <td>576</td>\n",
              "      <td>585</td>\n",
              "      <td>29</td>\n",
              "      <td>2076</td>\n",
              "      <td>397</td>\n",
              "      <td>122</td>\n",
              "      <td>12636</td>\n",
              "      <td>728</td>\n",
              "      <td>490</td>\n",
              "      <td>975</td>\n",
              "      <td>4267</td>\n",
              "      <td>2010</td>\n",
              "      <td>392</td>\n",
              "      <td>2119</td>\n",
              "      <td>351</td>\n",
              "      <td>12</td>\n",
              "      <td>901</td>\n",
              "      <td>276</td>\n",
              "      <td>1150</td>\n",
              "      <td>27</td>\n",
              "      <td>65</td>\n",
              "      <td>486</td>\n",
              "      <td>101</td>\n",
              "      <td>583</td>\n",
              "      <td>216</td>\n",
              "      <td>912</td>\n",
              "      <td>201</td>\n",
              "      <td>298</td>\n",
              "      <td>497</td>\n",
              "      <td>162</td>\n",
              "      <td>43</td>\n",
              "      <td>4466</td>\n",
              "      <td>110</td>\n",
              "      <td>87</td>\n",
              "      <td>447</td>\n",
              "      <td>...</td>\n",
              "      <td>125</td>\n",
              "      <td>42</td>\n",
              "      <td>1113</td>\n",
              "      <td>2113</td>\n",
              "      <td>762</td>\n",
              "      <td>118</td>\n",
              "      <td>94</td>\n",
              "      <td>3229</td>\n",
              "      <td>242</td>\n",
              "      <td>168</td>\n",
              "      <td>96</td>\n",
              "      <td>183</td>\n",
              "      <td>850</td>\n",
              "      <td>553</td>\n",
              "      <td>1033</td>\n",
              "      <td>507</td>\n",
              "      <td>508</td>\n",
              "      <td>14006</td>\n",
              "      <td>2589</td>\n",
              "      <td>191</td>\n",
              "      <td>1038</td>\n",
              "      <td>59</td>\n",
              "      <td>237</td>\n",
              "      <td>1482</td>\n",
              "      <td>254</td>\n",
              "      <td>1281</td>\n",
              "      <td>909</td>\n",
              "      <td>785</td>\n",
              "      <td>990</td>\n",
              "      <td>531</td>\n",
              "      <td>2233</td>\n",
              "      <td>22</td>\n",
              "      <td>1482</td>\n",
              "      <td>980</td>\n",
              "      <td>919</td>\n",
              "      <td>426</td>\n",
              "      <td>129</td>\n",
              "      <td>107</td>\n",
              "      <td>88</td>\n",
              "      <td>65</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>BM_36_516</th>\n",
              "      <td>2525</td>\n",
              "      <td>1258</td>\n",
              "      <td>130</td>\n",
              "      <td>6938</td>\n",
              "      <td>128</td>\n",
              "      <td>697</td>\n",
              "      <td>762</td>\n",
              "      <td>66</td>\n",
              "      <td>1303</td>\n",
              "      <td>480</td>\n",
              "      <td>168</td>\n",
              "      <td>12886</td>\n",
              "      <td>772</td>\n",
              "      <td>540</td>\n",
              "      <td>522</td>\n",
              "      <td>7075</td>\n",
              "      <td>2104</td>\n",
              "      <td>359</td>\n",
              "      <td>4588</td>\n",
              "      <td>326</td>\n",
              "      <td>57</td>\n",
              "      <td>1006</td>\n",
              "      <td>608</td>\n",
              "      <td>1018</td>\n",
              "      <td>44</td>\n",
              "      <td>27</td>\n",
              "      <td>357</td>\n",
              "      <td>306</td>\n",
              "      <td>751</td>\n",
              "      <td>227</td>\n",
              "      <td>927</td>\n",
              "      <td>383</td>\n",
              "      <td>256</td>\n",
              "      <td>901</td>\n",
              "      <td>170</td>\n",
              "      <td>39</td>\n",
              "      <td>4345</td>\n",
              "      <td>76</td>\n",
              "      <td>129</td>\n",
              "      <td>446</td>\n",
              "      <td>...</td>\n",
              "      <td>183</td>\n",
              "      <td>85</td>\n",
              "      <td>1283</td>\n",
              "      <td>1613</td>\n",
              "      <td>1283</td>\n",
              "      <td>151</td>\n",
              "      <td>163</td>\n",
              "      <td>3394</td>\n",
              "      <td>179</td>\n",
              "      <td>188</td>\n",
              "      <td>214</td>\n",
              "      <td>335</td>\n",
              "      <td>921</td>\n",
              "      <td>486</td>\n",
              "      <td>1309</td>\n",
              "      <td>566</td>\n",
              "      <td>587</td>\n",
              "      <td>17339</td>\n",
              "      <td>3131</td>\n",
              "      <td>263</td>\n",
              "      <td>1117</td>\n",
              "      <td>72</td>\n",
              "      <td>356</td>\n",
              "      <td>1601</td>\n",
              "      <td>182</td>\n",
              "      <td>1079</td>\n",
              "      <td>1022</td>\n",
              "      <td>693</td>\n",
              "      <td>1244</td>\n",
              "      <td>470</td>\n",
              "      <td>1385</td>\n",
              "      <td>35</td>\n",
              "      <td>1742</td>\n",
              "      <td>1001</td>\n",
              "      <td>920</td>\n",
              "      <td>287</td>\n",
              "      <td>117</td>\n",
              "      <td>280</td>\n",
              "      <td>157</td>\n",
              "      <td>59</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 284 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-cd93830e-ddfc-443f-adfb-b02432571ae9')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-cd93830e-ddfc-443f-adfb-b02432571ae9 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-cd93830e-ddfc-443f-adfb-b02432571ae9');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "              ENSG00000158186  ...  ENSG00000135587\n",
              "specimenID                     ...                 \n",
              "BM_36_517                3972  ...               88\n",
              "BM_36_308                2189  ...               61\n",
              "hB_RNA_10502             1704  ...               29\n",
              "hB_RNA_10617             3593  ...               65\n",
              "BM_36_516                2525  ...               59\n",
              "\n",
              "[5 rows x 284 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 125
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "msbb_phg_clusters = msbb_phg_clusters.set_index(\"specimenID\")\n",
        "msbb_phg_clusters"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 455
        },
        "id": "3qU9d7diOAdk",
        "outputId": "debc8068-c5dc-470c-e0f5-35f93729fc1c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-49ac788f-2695-4090-94a6-83880fb5b3b4\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>cluster</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>specimenID</th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>BM_36_1107</th>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>BM_36_278</th>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>BM_36_279</th>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>BM_36_281</th>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>BM_36_284</th>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>hB_RNA_11052</th>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>hB_RNA_11132</th>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>hB_RNA_11142</th>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>hB_RNA_12222</th>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>hB_RNA_12292</th>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>150 rows × 1 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-49ac788f-2695-4090-94a6-83880fb5b3b4')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-49ac788f-2695-4090-94a6-83880fb5b3b4 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-49ac788f-2695-4090-94a6-83880fb5b3b4');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "              cluster\n",
              "specimenID           \n",
              "BM_36_1107          1\n",
              "BM_36_278           1\n",
              "BM_36_279           1\n",
              "BM_36_281           1\n",
              "BM_36_284           2\n",
              "...               ...\n",
              "hB_RNA_11052        1\n",
              "hB_RNA_11132        1\n",
              "hB_RNA_11142        1\n",
              "hB_RNA_12222        1\n",
              "hB_RNA_12292        1\n",
              "\n",
              "[150 rows x 1 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 126
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GlXc5CzA8xI-"
      },
      "source": [
        "# Prepare data for Machine Learning\n",
        "\n",
        "First take 20% of randomly selected data from msbb_phg as test data."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_ING1tjcAoRz",
        "outputId": "c95e48a7-54c4-42eb-f47a-d8bbb66de413"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[(150, 284), (30, 284), (120, 284)]"
            ]
          },
          "metadata": {},
          "execution_count": 127
        }
      ],
      "source": [
        "#all_phg = msbb_phg.set_index(\"Ensembl ID\")\n",
        "#all_phg = msbb_phg.loc[:, msbb_phg.columns.isin(common_training)]\n",
        "#all_phg = all_phg.transpose()\n",
        "#all_phg.rename_axis(\"specimenID\", axis = 0, inplace=True)\n",
        "x_test = reduced_genes.sample(frac = 0.2)\n",
        "x_train = reduced_genes[~(reduced_genes.index.isin(set(x_test.index)))]\n",
        "[reduced_genes.shape, x_test.shape, x_train.shape]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "W0EbWpw_6gZK",
        "outputId": "fd0cab9b-8b5c-42ab-db7a-1e44b9041d13"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[(150, 1), (30, 1), (120, 1)]"
            ]
          },
          "metadata": {},
          "execution_count": 128
        }
      ],
      "source": [
        "y_train = msbb_phg_clusters\n",
        "y_test = y_train[y_train.index.isin(set(x_test.index))]\n",
        "y_train = y_train[y_train.index.isin(set(x_train.index))]\n",
        "[msbb_phg_clusters.shape, y_test.shape, y_train.shape]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "K-6vUWeV7sn_",
        "outputId": "268eee14-432b-4b7d-c7c3-bdd3a7b41b59"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().\n",
            "  return f(*args, **kwargs)\n"
          ]
        }
      ],
      "source": [
        "from sklearn.feature_selection import mutual_info_classif\n",
        "\n",
        "midf = pd.DataFrame({\n",
        "    \"feature\": x_train.columns,\n",
        "    \"mi\": mutual_info_classif(x_train, y_train)\n",
        "}).sort_values(\"mi\", ascending=False)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 282
        },
        "id": "dIttOltM868b",
        "outputId": "0ee4b83a-a357-43ef-8169-3a1a7f98cbcd"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[<matplotlib.lines.Line2D at 0x7f0be6141ed0>]"
            ]
          },
          "metadata": {},
          "execution_count": 131
        },
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAD4CAYAAADiry33AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAeeElEQVR4nO3de3hU1b3/8fc3kxu5AbkiJJBg4iWKF4yAN1SwCv5sqfVSbXvEHlu0rbX+tLXYnmM99vQ51fZobWtb6aPWaltErL9y1IocsLX1ggREICIQEUgQQgj3a27r98dsbJwGMjET9syez+t58mRm7zWT73LjJztr71nLnHOIiEhwpfhdgIiI9C8FvYhIwCnoRUQCTkEvIhJwCnoRkYBL9buASIWFha68vNzvMkREEsrixYu3OueKutsXd0FfXl5ObW2t32WIiCQUM1t/uH0auhERCTgFvYhIwEUV9GY2ycxWmVm9mU3vZv94M1tiZu1mdmWX7aeZ2etmVmdmy8zss7EsXkREetZj0JtZCHgImAxUA9eaWXVEsw3A9cDvI7bvA65zzp0ETAJ+YmaD+lq0iIhEL5qLsWOAeufcWgAzmwlMAd451MA5t87b19n1hc651V0ef2BmW4AiYEefKxcRkahEM3QzDGjo8rzR29YrZjYGSAfe62bfNDOrNbPa5ubm3r61iIgcwVG5GGtmxwBPAF90znVG7nfOzXDO1TjnaoqKur0NVEREPqZogn4jUNbleam3LSpmlgc8D3zXOfdG78qL3q4DbTwwbzVLGzQqJCLSVTRBvwioMrMKM0sHrgHmRPPmXvtngd8652Z//DJ75jrhwflrqF23rT9/jIhIwukx6J1z7cDNwFxgJTDLOVdnZveY2acAzOxMM2sErgIeNrM67+VXA+OB681sqfd1Wn90JG9AKqkpxra9rf3x9iIiCSuqKRCccy8AL0Rsu6vL40WEh3QiX/ck8GQfa4yKmTE4O52WPQp6EZGuAvXJ2ILsdFp0Ri8i8hGBCvr87HS27T3odxkiInElgEGvM3oRka4CFfQauhER+WeBCvr87Ax2H2intf2fPpMlIpK0ghX0OekAbN+ns3oRkUMCFfQF2eGg1y2WIiL/EKigz/eCXhdkRUT+IVBB/+EZvW6xFBH5UKCCXmf0IiL/LFBBPygrnVCKsXWPzuhFRA4JVNCHUoyinAyadinoRUQOCVTQA5QMzKRp1wG/yxARiRuBC/oheRls3qmgFxE5JHBBX5KnM3oRka4CGfS7DrSzv7XD71JEROJC4IJ+SF4mAJt1Vi8iAgQx6Ad6Qa9xehERIIBBX+Kd0W/ZraAXEYEABr3O6EVEPipwQZ+TkUpORiqbFPQiIkAAgx6gLD+Lhm37/C5DRCQuBDLoR+Rnsa5lr99liIjEhWAGfWEWDdv209Hp/C5FRMR3UQW9mU0ys1VmVm9m07vZP97MlphZu5ldGbFvqpmt8b6mxqrwIxmRn01rR6fupRcRIYqgN7MQ8BAwGagGrjWz6ohmG4Drgd9HvDYf+B4wFhgDfM/MBve97CMrL8gCYP1WDd+IiERzRj8GqHfOrXXOtQIzgSldGzjn1jnnlgGdEa+9BJjnnNvmnNsOzAMmxaDuIxpRmA3AuhZdkBURiSbohwENXZ43etuiEdVrzWyamdWaWW1zc3OUb314Q/IySQ+lsH6bzuhFROLiYqxzboZzrsY5V1NUVNTn9wulGBWF2azctDsG1YmIJLZogn4jUNbleam3LRp9eW2fjBuZz6L3t3GwXbNYikhyiyboFwFVZlZhZunANcCcKN9/LnCxmQ32LsJe7G3rd+dUFrK/rYO3Nuw4Gj9ORCRu9Rj0zrl24GbCAb0SmOWcqzOze8zsUwBmdqaZNQJXAQ+bWZ332m3A9wn/slgE3ONt63fjji0glGK8Wr/1aPw4EZG4Zc7F14eKampqXG1tbUze6+qHX2dDyz7m3Tae3My0mLyniEg8MrPFzrma7vbFxcXY/nLn5BNo2n2A7z/3DvH2C01E5GgJdNCfPnwwXzn/WGbVNvLjl1b5XY6IiC8CHfQA37rkeC46sYSZbzb03FhEJIACH/RmxklD82jZ20pbR+QHd0VEgi/wQQ9QnJcBwNY9B32uRETk6EuOoM/11pHdpaAXkeSTJEEfPqPfsltBLyLJJzmCPu9Q0Gt+ehFJPkkR9IU5GZhp6EZEklNSBH1aKIX8rHQN3YhIUkqKoAcoys2gWUM3IpKEkiboi/MydUYvIkkpeYI+N4MmLRYuIkkoaYL+hCG5NO06yLubd/ldiojIUZU0QX/F6FLSU1N48o31fpciInJUJU3QD85O55OnDOXZJRvZfaDN73JERI6apAl6gOvOGsHe1g7+uOSoLFsrIhIXkiroTy0bxKmlA3n89XW0aIIzEUkSSRX0AF+54FjWNu/lrB8u4LZZS3mpbrNWnxKRQEu6oJ908jH8723j+WxNGS/VNTHticU8/Mpav8sSEek3SRf0AJXFuXz/0yfz1l2f4KITi/np/DWsbd7jd1kiIv0iKYP+kLRQCndddhIhMyb95G88t+wDv0sSEYm5pA56gOEFWcy77Xyqh+bx3WdX0KxpEkQkYJI+6AGGDMzkx1edyp6D7fzmtff9LkdEJKYU9J7K4hyGDsqkcft+v0sREYmpqILezCaZ2Sozqzez6d3szzCzp7z9C82s3NueZmaPm9lyM1tpZnfGtvzYKs7N1OIkIhI4PQa9mYWAh4DJQDVwrZlVRzS7AdjunKsEHgDu9bZfBWQ450YBZwA3HvolEI+KczO03KCIBE40Z/RjgHrn3FrnXCswE5gS0WYK8Lj3eDYw0cwMcEC2maUCA4BWIG6njwwHvc7oRSRYogn6YUBDl+eN3rZu2zjn2oGdQAHh0N8LbAI2AD92zm2L/AFmNs3Mas2strm5udediJXivEx2H2jnQFuHbzWIiMRaf1+MHQN0AEOBCuB2MxsZ2cg5N8M5V+OcqykqKurnkg6vKDcDQLdYikigRBP0G4GyLs9LvW3dtvGGaQYCLcDngBedc23OuS3Aq0BNX4vuL4eCXuP0IhIk0QT9IqDKzCrMLB24BpgT0WYOMNV7fCWwwIVnCtsATAAws2xgHPBuLArvD8WHgl533ohIgPQY9N6Y+83AXGAlMMs5V2dm95jZp7xmjwAFZlYP3AYcugXzISDHzOoI/8J4zDm3LNadiJXi3EwAXZAVkUBJjaaRc+4F4IWIbXd1eXyA8K2Uka/b0932eFWQnU4oxTR0IyKBok/GdpGSYhTlZLB5p87oRSQ4FPQRKgqzeU9TFotIgCjoIxw/JJc1Tbvp7NSqUyISDAr6CMcPyWVvawcbd2hyMxEJBgV9hONKcgFY3bTb50pERGJDQR/huJIcAN7drKAXkWBQ0EfIzUxj2KABOqMXkcBQ0HdjZFE261r2+V2GiEhMKOi7UTo4i8ZtCnoRCQYFfTfK8gfQsreVvQfb/S5FRKTPFPTdGJ6fBUDDdp3Vi0jiU9B3o2ywF/TbdC+9iCQ+BX03yrwz+g0apxeRAFDQd2NwVhrZ6SEaFPQiEgAK+m6YGWX5WQp6EQkEBf1hjBo2kNfea9Hc9CKS8BT0h/G1Cytp6+jkZ/Pr/S5FRKRPolphKhmVF2bzhXEj+M1r66gszuGU0oFkpadSmJNOQU6G3+WJiERNQX8Ed156Ais27uR7c+o+3BZKMf76rQso9W7BFBGJdwr6I8hIDfHkl8aytGEH+1s7WN+yl7v/5x0Wr9+uoBeRhKEx+h5kpoUYN7KAC08o5gvjRpCZlsKyxp1+lyUiEjUFfS+khlI4aehAljXu8LsUEZGoKeh7adSwgazYuIsOrSkrIglCQd9Lp5YNZH9bB+9u3uV3KSIiUYkq6M1skpmtMrN6M5vezf4MM3vK27/QzMq77DvFzF43szozW25mmbEr/+g7t7KI1BTjj0s2+l2KiEhUegx6MwsBDwGTgWrgWjOrjmh2A7DdOVcJPADc6702FXgSuMk5dxJwAdAWs+p9UJSbwSUnD2H24kYOtHX4XY6ISI+iOaMfA9Q759Y651qBmcCUiDZTgMe9x7OBiWZmwMXAMufc2wDOuRbnXMKn49Szytm5v43rHn2TTTs1lbGIxLdogn4Y0NDleaO3rds2zrl2YCdQABwHODOba2ZLzOyO7n6AmU0zs1ozq21ubu5tH466MRX53H/1qazYuJNJP/kb059ZxhotJi4icaq/L8amAucCn/e+X25mEyMbOedmOOdqnHM1RUVF/VxSbHxmdCnP33Iepw8fxOzFjTz22jq/SxIR6VY0Qb8RKOvyvNTb1m0bb1x+INBC+Oz/FefcVufcPuAFYHRfi44XFYXZ/OaLYzh9+CDqm/b4XY6ISLeiCfpFQJWZVZhZOnANMCeizRxgqvf4SmCBc84Bc4FRZpbl/QI4H3gnNqXHj8riXFZv2U24yyIi8aXHoPfG3G8mHNorgVnOuTozu8fMPuU1ewQoMLN64DZguvfa7cD9hH9ZLAWWOOeej303/FVVnMOOfW1s3dPqdykiIv8kqknNnHMvEB526brtri6PDwBXHea1TxK+xTKwjivJBWBN026KcjWFsYjEF30yNgaqSnIAWLNF4/QiEn8U9DFQnJtBXmYqSxs02ZmIxB8FfQyYGZ88dSjPL9tE0y6tMSsi8UVBHyPTxo+kvbOTR1993+9SREQ+QkEfIyMKsvlEdQnPLN5Ie0en3+WIiHxIQR9Dl58+jK17DvLqey1+lyIi8iEFfQxdcHwxeZmp/GHhBr9LERH5kII+hjLTQlx/TgUv1m3m0b9rrF5E4oOCPsa+MbGKi04s4Z7n3uGbT7/N3oPtfpckIklOQR9joRTjV18YzdcnVPLMkkYu/8Wr7G9N+Cn4RSSBKej7QWoohdsvPp5HptawumkP9819l537EnphLRFJYAr6fjThhBKurinlsVfXMfo/53H/vNV0dGqGSxE5uqKa1Ew+vh9cPoqJJ5bw5+Wb+On8NTjnuP3i4/0uS0SSiIK+n6WFUrjkpCFcXF1CemoKP1tQz5+WfsDXJ1Ry5RmlhJfWFRHpPwr6o8TMuGfKyYwsyuGlus18a/YyXnuvhc+NHU71MXlkZ+hQiEj/sHhbFammpsbV1tb6XUa/6uh0/HxBPQ/OX02ng2GDBnD/1acydmSB36WJSIIys8XOuZru9ulirA9CKcY3Lqpi/u0X8KsvjCYtZFzz6zd4TBOiiUg/UND7qKIwm0knH8Pzt5zHWSML+On8NbRpQjQRiTEFfRzIzkhl6tnlbN/XxhtrNSGaiMSWgj5OnH9cEdnpIZ5ftsnvUkQkYBT0cSIzLcTEE0uY904TnfpQlYjEkII+jlx4QhEte1tZ8cFOv0sRkQBR0MeR8VVFmMFfVjX7XYqIBIiCPo4U5GRwyrCB/HFJI0sbdvhdjogEhII+znz1wkq272vj6odfp2nXAb/LEZEAiCrozWySma0ys3ozm97N/gwze8rbv9DMyiP2DzezPWb2zdiUHVyXnDSEOTefQ3tHJ49olSoRiYEeg97MQsBDwGSgGrjWzKojmt0AbHfOVQIPAPdG7L8f+HPfy00OIwqyueyUocx4ZS3n/+hlNu7Y73dJIpLAojmjHwPUO+fWOudagZnAlIg2U4DHvcezgYnmTctoZp8G3gfqYlNycrhnykl865Ljadp1gP+Yo/90IvLxRRP0w4CGLs8bvW3dtnHOtQM7gQIzywG+DfzHkX6AmU0zs1ozq21u1h0nAIOy0vnahZXcMrGKl95p4hsz36J+y26/yxKRBNTfc+PeDTzgnNtzpHnXnXMzgBkQnr2yn2tKKDeOP5bW9k5+7s1jP2rYQL48fiSXjTqGlBTNZS8iPYvmjH4jUNbleam3rds2ZpYKDARagLHAfWa2DrgV+I6Z3dzHmpNKKMW49aLjeO3OCfz7ZdUcbO/glj+8xa1PLaVdE6CJSBSiOaNfBFSZWQXhQL8G+FxEmznAVOB14EpggQtPdH/eoQZmdjewxzn38xjUnXSKczO54dwKrj+7nF/+pZ4fv7Sa4twM/u2yyOviIiIf1WPQO+favbPwuUAIeNQ5V2dm9wC1zrk5wCPAE2ZWD2wj/MtA+kEoxbh5QhWbdh7g0Vff54LjizmnskBLEorIYWmFqQS160AbF9//Cpt3HSAtZAwckM5xJTl8fuwIxh9XSG5mmt8lishRdKQVphT0CWzrnoMsWLmF91v2smNfK6+s3srGHfvJzUjlwWtPY8IJJX6XKCJHyZGCXitSJ7DCnAyuPvMf18nbOzqpXb+d7z/3Dl/+7WJm3XgWZ4wY7GOFIhIPNNdNgKSGUhg3soA/TBvHkLxMbp+1lANtHX6XJSI+U9AHUF5mGv/1mVGsa9nH/7z9gd/liIjPFPQBdV5VISOLsnlqUUPPjUUk0BT0AWVmXHNmGbXrt7Nio1asEklmCvoAu7qmjPzsdP7t/62gQ+vQiiQtBX2ADcpK567LqlnasIPfLVzvdzki4hMFfcBNOW0o51UVct+Lq5hbt5kNLfuIt89OiEj/UtAHnJnxg0+PIj01hRufWMz4H73MTU8uprVdE6KJJAsFfRIYXpDFa9Mn8PRNZ3HLhErm1jXxf59aqjN7kSShT8Ymicy0EGeW53NmeT5ZGan88M/vUvxcBt+edAKZaSG/yxORfqSgT0I3jh9J4/Z9PPbqOja07OOR68/0uyQR6UcK+iRkZvznp0eRm5nGr/76Hlt2H6A4N9PvskSkn2iMPoldfvownIO5Kzb7XYqI9CMFfRI7riSXyuIcnlmykU59oEoksBT0Se5L51awtGEHD85fo7twRAJKY/RJ7rNnlrHw/W08OH8Ni9Zt49yqQm4afywpKVqaUCQodEaf5MyMH191KndMOp7G7fu578VVzFvZ5HdZIhJDCnohlGJ89YJKFtx+PqWDBzDjlbV+lyQiMaSglw+lhlK44dwKFq/fzqrNu/0uR0RiREEvH3HRieEFxV9/b6vPlYhIrCjo5SPK8rMoHTyA19e2+F2KiMSIgl7+yVkjC1j4/jbdWy8SEFEFvZlNMrNVZlZvZtO72Z9hZk95+xeaWbm3/RNmttjMlnvfJ8S2fOkP40YWsGNfG08v1nqzIkHQY9CbWQh4CJgMVAPXmll1RLMbgO3OuUrgAeBeb/tW4JPOuVHAVOCJWBUu/WfyqCGMrcjn288sZ9pva9m5r83vkkSkD6I5ox8D1Dvn1jrnWoGZwJSINlOAx73Hs4GJZmbOubeccx942+uAAWaWEYvCpf9kpafy5JfG8q1LjmfeyiYe+btutxRJZNEE/TCg69/wjd62bts459qBnUBBRJsrgCXOuYORP8DMpplZrZnVNjc3R1u79KO0UApfu7CSsRX5PLd8k6ZHEElgR+VirJmdRHg458bu9jvnZjjnapxzNUVFRUejJInSZacMZW3zXt7ZtMvvUkTkY4om6DcCZV2el3rbum1jZqnAQKDFe14KPAtc55x7r68Fy9E1+eQhZKalcP1ji/jFX+pp2LbP75JEpJeiCfpFQJWZVZhZOnANMCeizRzCF1sBrgQWOOecmQ0CngemO+dejVXRcvQU5GTw9I1nU16QxX0vrmL8j17mF3+p97ssEemFHoPeG3O/GZgLrARmOefqzOweM/uU1+wRoMDM6oHbgEO3YN4MVAJ3mdlS76s45r2QfjWqdCBP33Q2r06fwNnHFvDwX9fS2t7pd1kiEiWLt4tsNTU1rra21u8y5DBeXrWFLz62iF9fV8Mnqkv8LkdEPGa22DlX090+fTJWeuW8ykIKstOZ+eYGv0sRkSgp6KVXUkMp/Ou5Fcx/dwuzavXJWZFEoKCXXrvp/GMZU57PHbOXcfust/0uR0R6oKCXXgulGL/51zP5zOnDeGZJIzv3a4oEkXimoJePJSs9lctHhz8gvaxxh8/ViMiRKOjlYzu1bBBm8NYGBb1IPFPQy8eWl5lGVXEOb23Y7ncpInIECnrpk9PLBlO7frumRhCJY6l+FyCJ7V/OGsGLdZuZ+N9/JW9AGlnpIbLSQ4weMZjbP3EcBTmalVrEbwp66ZOThw3k2a+ezVOLGth1oJ39re3s2N/G07UNbN/byi+/cIbfJYokPQW99NnIohzuvPTEj2x76OV6fjR3Fd95djlfPLucqpJcn6oTEQW99IsvnzeSNU27eWZxIzPf3MD/OWUoF51YTEleJsPzsxg6aIDfJYokDU1qJv1q295WfvFyPU8v/scHq1IMPjO6lB9+ZhSpId0PIBILR5rUTGf00q/ys9P5t8uquWPSCbzXvIdte1tZ8O4WHvn7++w+0MbUs8s5+9hCv8sUCTQFvRwV6akpnHhMHgDnVBYyaEAaD85fw9y6Jr4/5ST+5axyfwsUCTD93Sy++PrEKpbffQkXnVjCv/+pjt8v3MDOfZozR6Q/aIxefHWgrYMv/7aWv63ZCkDNiMGMG1nAF88p1z34Ir2gMXqJW5lpIX59XQ3PL9vExh37eXHFZn7+cj0ZqSl8fWKV3+WJBIKCXnyXmRbiijNKAbhlYhUX3f9X3taMmCIxozF6iTunlg5iacNO4m1YUSRRKegl7pxWNpCtew7ywc4DfpciEggKeok7p5YNAuDva5p9rkQkGBT0EndOGJJHSV4G335mOXfMfpuOTg3hiPSFgl7iTnpqCs/fch7Txo9kVm0jV/zyNebWbfa7LJGEpaCXuFSYk8F3Lj2RH1x+Mrv2t3HjE4v56fw1fpclkpCiCnozm2Rmq8ys3symd7M/w8ye8vYvNLPyLvvu9LavMrNLYle6JIPPjx3Bi7eO5/LTh3H/vNW8sHyT3yWJJJweg97MQsBDwGSgGrjWzKojmt0AbHfOVQIPAPd6r60GrgFOAiYBv/DeTyRq6akp3HvFKZxWNohbZy5lVm2Dbr0U6YVoPjA1Bqh3zq0FMLOZwBTgnS5tpgB3e49nAz83M/O2z3TOHQTeN7N67/1ej035kizSU1N47Poz+crvFnPH7GX8ZN5qsjP0eT8JlhOOyeNn154e8/eN5v+UYUBDl+eNwNjDtXHOtZvZTqDA2/5GxGuHRf4AM5sGTAMYPnx4tLVLkhmcnc6TN4zlT0s/YMGqLTqrl8ApG9w/C/LExSmRc24GMAPCk5r5XI7EsdRQClecUfrhlAki0rNoLsZuBMq6PC/1tnXbxsxSgYFAS5SvFRGRfhRN0C8CqsyswszSCV9cnRPRZg4w1Xt8JbDAhf+ungNc492VUwFUAW/GpnQREYlGj0M33pj7zcBcIAQ86pyrM7N7gFrn3BzgEeAJ72LrNsK/DPDazSJ84bYd+JpzrqOf+iIiIt3QwiMiIgFwpIVH9MlYEZGAU9CLiAScgl5EJOAU9CIiARd3F2PNrBlY34e3KAS2xqiceBHEPoH6lWjUr/g2wjlX1N2OuAv6vjKz2sNdeU5UQewTqF+JRv1KXBq6EREJOAW9iEjABTHoZ/hdQD8IYp9A/Uo06leCCtwYvYiIfFQQz+hFRKQLBb2ISMAFJuh7WsA8kZjZOjNbbmZLzazW25ZvZvPMbI33fbDfdfbEzB41sy1mtqLLtm77YWE/9Y7fMjMb7V/lR3aYft1tZhu9Y7bUzC7tsu9Or1+rzOwSf6o+MjMrM7OXzewdM6szs2942xP6eB2hXwl9vHrNOZfwX4SnT34PGAmkA28D1X7X1Yf+rAMKI7bdB0z3Hk8H7vW7zij6MR4YDazoqR/ApcCfAQPGAQv9rr+X/bob+GY3bau9f48ZQIX37zTkdx+6qfMYYLT3OBdY7dWe0MfrCP1K6OPV26+gnNF/uIC5c64VOLSAeZBMAR73Hj8OfNrHWqLinHuF8PoEXR2uH1OA37qwN4BBZnbM0am0dw7Tr8OZAsx0zh10zr0P1BP+9xpXnHObnHNLvMe7gZWE13dO6ON1hH4dTkIcr94KStB3t4D5kQ5mvHPAS2a22Fs4HaDEObfJe7wZKPGntD47XD+CcAxv9oYxHu0ytJZw/TKzcuB0YCEBOl4R/YKAHK9oBCXog+Zc59xoYDLwNTMb33WnC/+NmfD3xQalH55fAscCpwGbgP/2t5yPx8xygGeAW51zu7ruS+Tj1U2/AnG8ohWUoA/UIuTOuY3e9y3As4T/dGw69Kex932LfxX2yeH6kdDH0DnX5JzrcM51Ar/mH3/uJ0y/zCyNcBj+zjn3R29zwh+v7voVhOPVG0EJ+mgWME8IZpZtZrmHHgMXAyv46ALsU4E/+VNhnx2uH3OA67y7OcYBO7sMGcS9iPHpywkfMwj36xozyzCzCqAKePNo19cTMzPCaz+vdM7d32VXQh+vw/Ur0Y9Xr/l9NThWX4TvAlhN+Cr5d/2upw/9GEn4qv/bQN2hvgAFwHxgDfC/QL7ftUbRlz8Q/rO4jfBY5w2H6wfhuzce8o7fcqDG7/p72a8nvLqXEQ6LY7q0/67Xr1XAZL/rP0yfziU8LLMMWOp9XZrox+sI/Uro49XbL02BICIScEEZuhERkcNQ0IuIBJyCXkQk4BT0IiIBp6AXEQk4Bb2ISMAp6EVEAu7/A3qC2K20FQhfAAAAAElFTkSuQmCC\n",
            "text/plain": [
              "<Figure size 432x288 with 1 Axes>"
            ]
          },
          "metadata": {
            "needs_background": "light"
          }
        }
      ],
      "source": [
        "import matplotlib.pyplot as plt\n",
        "import numpy as np\n",
        "\n",
        "plt.plot(np.arange(284), midf.mi.iloc[:284])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3ZcWqflX9BLk",
        "outputId": "ccaa1659-561e-4a7e-8fb0-7e656baad71a"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "107"
            ]
          },
          "metadata": {},
          "execution_count": 133
        }
      ],
      "source": [
        "selected_genes = midf.loc[midf.mi >= 0.005, \"feature\"]\n",
        "len(selected_genes)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "l8jcB9Hp_bVl",
        "outputId": "4628a69e-aea8-4dbf-f14d-c7ec5d5a5c7a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning: path already exists! This predictor may overwrite an existing predictor! path=\"good_quality_cluster\"\n",
            "Presets specified: ['good_quality_faster_inference_only_refit']\n",
            "Beginning AutoGluon training ...\n",
            "AutoGluon will save models to \"good_quality_cluster/\"\n",
            "AutoGluon Version:  0.3.1\n",
            "Train Data Rows:    120\n",
            "Train Data Columns: 284\n",
            "Preprocessing data ...\n",
            "AutoGluon infers your prediction problem is: 'multiclass' (because dtype of label-column == int, but few unique label-values observed).\n",
            "\t3 unique label values:  [1, 3, 2]\n",
            "\tIf 'multiclass' is not the correct problem_type, please manually specify the problem_type argument in fit() (You may specify problem_type as one of: ['binary', 'multiclass', 'regression'])\n",
            "Train Data Class Count: 3\n",
            "Using Feature Generators to preprocess the data ...\n",
            "Fitting AutoMLPipelineFeatureGenerator...\n",
            "\tAvailable Memory:                    7193.11 MB\n",
            "\tTrain Data (Original)  Memory Usage: 0.27 MB (0.0% of available memory)\n",
            "\tInferring data type of each feature based on column values. Set feature_metadata_in to manually specify special dtypes of the features.\n",
            "\tStage 1 Generators:\n",
            "\t\tFitting AsTypeFeatureGenerator...\n",
            "\tStage 2 Generators:\n",
            "\t\tFitting FillNaFeatureGenerator...\n",
            "\tStage 3 Generators:\n",
            "\t\tFitting IdentityFeatureGenerator...\n",
            "\tStage 4 Generators:\n",
            "\t\tFitting DropUniqueFeatureGenerator...\n",
            "\tTypes of features in original data (raw dtype, special dtypes):\n",
            "\t\t('int', []) : 284 | ['ENSG00000158186', 'ENSG00000135766', 'ENSG00000179943', 'ENSG00000171603', 'ENSG00000269190', ...]\n",
            "\tTypes of features in processed data (raw dtype, special dtypes):\n",
            "\t\t('int', []) : 284 | ['ENSG00000158186', 'ENSG00000135766', 'ENSG00000179943', 'ENSG00000171603', 'ENSG00000269190', ...]\n",
            "\t0.6s = Fit runtime\n",
            "\t284 features in original data used to generate 284 features in processed data.\n",
            "\tTrain Data (Processed) Memory Usage: 0.27 MB (0.0% of available memory)\n",
            "Data preprocessing and feature engineering runtime = 0.84s ...\n",
            "AutoGluon will gauge predictive performance using evaluation metric: 'accuracy'\n",
            "\tTo change this, specify the eval_metric argument of fit()\n",
            "Fitting 11 L1 models ...\n",
            "Fitting model: NeuralNetFastAI_BAG_L1 ...\n",
            "No improvement since epoch 8: early stopping\n",
            "No improvement since epoch 6: early stopping\n",
            "No improvement since epoch 6: early stopping\n",
            "\t0.575\t = Validation score   (accuracy)\n",
            "\t47.03s\t = Training   runtime\n",
            "\t1.29s\t = Validation runtime\n",
            "Fitting model: LightGBMXT_BAG_L1 ...\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "\t0.5417\t = Validation score   (accuracy)\n",
            "\t2.21s\t = Training   runtime\n",
            "\t0.04s\t = Validation runtime\n",
            "Fitting model: LightGBM_BAG_L1 ...\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "\t0.5333\t = Validation score   (accuracy)\n",
            "\t2.18s\t = Training   runtime\n",
            "\t0.03s\t = Validation runtime\n",
            "Fitting model: RandomForestGini_BAG_L1 ...\n",
            "\t0.45\t = Validation score   (accuracy)\n",
            "\t1.06s\t = Training   runtime\n",
            "\t0.11s\t = Validation runtime\n",
            "Fitting model: RandomForestEntr_BAG_L1 ...\n",
            "\t0.45\t = Validation score   (accuracy)\n",
            "\t1.25s\t = Training   runtime\n",
            "\t0.11s\t = Validation runtime\n",
            "Fitting model: CatBoost_BAG_L1 ...\n",
            "\t0.5917\t = Validation score   (accuracy)\n",
            "\t52.56s\t = Training   runtime\n",
            "\t0.08s\t = Validation runtime\n",
            "Fitting model: ExtraTreesGini_BAG_L1 ...\n",
            "\t0.5083\t = Validation score   (accuracy)\n",
            "\t1.07s\t = Training   runtime\n",
            "\t0.11s\t = Validation runtime\n",
            "Fitting model: ExtraTreesEntr_BAG_L1 ...\n",
            "\t0.4667\t = Validation score   (accuracy)\n",
            "\t1.07s\t = Training   runtime\n",
            "\t0.11s\t = Validation runtime\n",
            "Fitting model: XGBoost_BAG_L1 ...\n",
            "\t0.4333\t = Validation score   (accuracy)\n",
            "\t9.44s\t = Training   runtime\n",
            "\t0.03s\t = Validation runtime\n",
            "Fitting model: NeuralNetMXNet_BAG_L1 ...\n",
            "\tWarning: Exception caused NeuralNetMXNet_BAG_L1 to fail during training (ImportError)... Skipping this model.\n",
            "\t\tUnable to import dependency mxnet. A quick tip is to install via `pip install mxnet --upgrade`, or `pip install mxnet_cu101 --upgrade`\n",
            "Fitting model: LightGBMLarge_BAG_L1 ...\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "\t0.525\t = Validation score   (accuracy)\n",
            "\t8.71s\t = Training   runtime\n",
            "\t0.03s\t = Validation runtime\n",
            "Fitting model: WeightedEnsemble_L2 ...\n",
            "\t0.6\t = Validation score   (accuracy)\n",
            "\t0.29s\t = Training   runtime\n",
            "\t0.0s\t = Validation runtime\n",
            "AutoGluon training complete, total runtime = 130.73s ...\n",
            "Fitting 1 L1 models ...\n",
            "Fitting model: NeuralNetFastAI_BAG_L1_FULL ...\n",
            "\tStopping at the best epoch learned earlier - 8.\n",
            "\t1.42s\t = Training   runtime\n",
            "Fitting 1 L1 models ...\n",
            "Fitting model: LightGBMXT_BAG_L1_FULL ...\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "\t0.24s\t = Training   runtime\n",
            "Fitting 1 L1 models ...\n",
            "Fitting model: LightGBM_BAG_L1_FULL ...\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "\t0.22s\t = Training   runtime\n",
            "Fitting 1 L1 models ...\n",
            "Fitting model: RandomForestGini_BAG_L1_FULL ...\n",
            "\t0.45\t = Validation score   (accuracy)\n",
            "\t1.04s\t = Training   runtime\n",
            "\t0.11s\t = Validation runtime\n",
            "Fitting 1 L1 models ...\n",
            "Fitting model: RandomForestEntr_BAG_L1_FULL ...\n",
            "\t0.45\t = Validation score   (accuracy)\n",
            "\t1.14s\t = Training   runtime\n",
            "\t0.11s\t = Validation runtime\n",
            "Fitting 1 L1 models ...\n",
            "Fitting model: CatBoost_BAG_L1_FULL ...\n",
            "\t0.49s\t = Training   runtime\n",
            "Fitting 1 L1 models ...\n",
            "Fitting model: ExtraTreesGini_BAG_L1_FULL ...\n",
            "\t0.5083\t = Validation score   (accuracy)\n",
            "\t1.07s\t = Training   runtime\n",
            "\t0.12s\t = Validation runtime\n",
            "Fitting 1 L1 models ...\n",
            "Fitting model: ExtraTreesEntr_BAG_L1_FULL ...\n",
            "\t0.4667\t = Validation score   (accuracy)\n",
            "\t1.09s\t = Training   runtime\n",
            "\t0.11s\t = Validation runtime\n",
            "Fitting 1 L1 models ...\n",
            "Fitting model: XGBoost_BAG_L1_FULL ...\n",
            "\t1.62s\t = Training   runtime\n",
            "Fitting 1 L1 models ...\n",
            "Fitting model: LightGBMLarge_BAG_L1_FULL ...\n",
            "/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:239: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.\n",
            "  _log_warning(\"'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. \"\n",
            "\t0.35s\t = Training   runtime\n",
            "Fitting model: WeightedEnsemble_L2_FULL ...\n",
            "\t0.6\t = Validation score   (accuracy)\n",
            "\t0.0s\t = Training   runtime\n",
            "\t0.0s\t = Validation runtime\n",
            "TabularPredictor saved. To load, use: predictor = TabularPredictor.load(\"good_quality_cluster/\")\n"
          ]
        }
      ],
      "source": [
        "from autogluon.tabular import TabularPredictor\n",
        "train_data = x_train.copy().loc[:, selected_genes].copy()\n",
        "train_data[\"cluster\"] = y_train.values\n",
        "\n",
        "predictor = TabularPredictor(label=\"cluster\", path=\"good_quality_cluster\").fit( \n",
        "    train_data=train_data,\n",
        "    presets=\"good_quality_faster_inference_only_refit\",\n",
        ")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sHFzIsnz_nN1",
        "outputId": "4e8d86e2-c06a-47ca-d985-28b649ed6ffa"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "*** Summary of fit() ***\n",
            "Estimated performance of each model:\n",
            "                           model  score_val  pred_time_val   fit_time  pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  fit_order\n",
            "0                CatBoost_BAG_L1   0.600000       0.037702  23.495608                0.037702          23.495608            1      False          6\n",
            "1            WeightedEnsemble_L2   0.600000       0.038233  23.789553                0.000531           0.293945            2      False         11\n",
            "2              LightGBMXT_BAG_L1   0.558333       0.023321   1.432673                0.023321           1.432673            1      False          2\n",
            "3           LightGBMLarge_BAG_L1   0.541667       0.025970   5.933805                0.025970           5.933805            1      False         10\n",
            "4                LightGBM_BAG_L1   0.533333       0.020676   1.684660                0.020676           1.684660            1      False          3\n",
            "5                 XGBoost_BAG_L1   0.525000       0.036073   4.371852                0.036073           4.371852            1      False          9\n",
            "6     ExtraTreesEntr_BAG_L1_FULL   0.508333       0.110222   1.004183                0.110222           1.004183            1       True         19\n",
            "7          ExtraTreesEntr_BAG_L1   0.508333       0.113128   0.900875                0.113128           0.900875            1      False          8\n",
            "8         NeuralNetFastAI_BAG_L1   0.500000       0.192538   8.605374                0.192538           8.605374            1      False          1\n",
            "9        RandomForestGini_BAG_L1   0.466667       0.109918   0.988833                0.109918           0.988833            1      False          4\n",
            "10  RandomForestGini_BAG_L1_FULL   0.466667       0.114366   0.890958                0.114366           0.890958            1       True         15\n",
            "11         ExtraTreesGini_BAG_L1   0.458333       0.103192   0.898167                0.103192           0.898167            1      False          7\n",
            "12    ExtraTreesGini_BAG_L1_FULL   0.458333       0.118871   0.992579                0.118871           0.992579            1       True         18\n",
            "13       RandomForestEntr_BAG_L1   0.425000       0.105163   1.089415                0.105163           1.089415            1      False          5\n",
            "14  RandomForestEntr_BAG_L1_FULL   0.425000       0.125314   1.102144                0.125314           1.102144            1       True         16\n",
            "15           XGBoost_BAG_L1_FULL        NaN            NaN   0.549147                     NaN           0.549147            1       True         20\n",
            "16      WeightedEnsemble_L2_FULL        NaN            NaN   0.828163                0.000618           0.000318            2       True         22\n",
            "17   NeuralNetFastAI_BAG_L1_FULL        NaN            NaN   0.890719                     NaN           0.890719            1       True         12\n",
            "18          LightGBM_BAG_L1_FULL        NaN            NaN   0.183733                     NaN           0.183733            1       True         14\n",
            "19        LightGBMXT_BAG_L1_FULL        NaN            NaN   0.204114                     NaN           0.204114            1       True         13\n",
            "20     LightGBMLarge_BAG_L1_FULL        NaN            NaN   0.232890                     NaN           0.232890            1       True         21\n",
            "21          CatBoost_BAG_L1_FULL        NaN            NaN   0.827846                     NaN           0.827846            1       True         17\n",
            "Number of models trained: 22\n",
            "Types of models trained:\n",
            "{'StackerEnsembleModel_LGB', 'StackerEnsembleModel_XT', 'StackerEnsembleModel_NNFastAiTabular', 'StackerEnsembleModel_CatBoost', 'StackerEnsembleModel_XGBoost', 'StackerEnsembleModel_RF', 'WeightedEnsembleModel'}\n",
            "Bagging used: True  (with 5 folds)\n",
            "Multi-layer stack-ensembling used: False \n",
            "Feature Metadata (Processed):\n",
            "(raw dtype, special dtypes):\n",
            "('int', []) : 107 | ['ENSG00000174238', 'ENSG00000187990', 'ENSG00000166326', 'ENSG00000102103', 'ENSG00000141655', ...]\n",
            "Plot summary of models saved to file: good_quality_cluster/SummaryOfModels.html\n",
            "*** End of fit() summary ***\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'leaderboard':                            model  score_val  ...  can_infer  fit_order\n",
              " 0                CatBoost_BAG_L1   0.600000  ...      False          6\n",
              " 1            WeightedEnsemble_L2   0.600000  ...      False         11\n",
              " 2              LightGBMXT_BAG_L1   0.558333  ...      False          2\n",
              " 3           LightGBMLarge_BAG_L1   0.541667  ...      False         10\n",
              " 4                LightGBM_BAG_L1   0.533333  ...      False          3\n",
              " 5                 XGBoost_BAG_L1   0.525000  ...      False          9\n",
              " 6     ExtraTreesEntr_BAG_L1_FULL   0.508333  ...       True         19\n",
              " 7          ExtraTreesEntr_BAG_L1   0.508333  ...      False          8\n",
              " 8         NeuralNetFastAI_BAG_L1   0.500000  ...      False          1\n",
              " 9        RandomForestGini_BAG_L1   0.466667  ...      False          4\n",
              " 10  RandomForestGini_BAG_L1_FULL   0.466667  ...       True         15\n",
              " 11         ExtraTreesGini_BAG_L1   0.458333  ...      False          7\n",
              " 12    ExtraTreesGini_BAG_L1_FULL   0.458333  ...       True         18\n",
              " 13       RandomForestEntr_BAG_L1   0.425000  ...      False          5\n",
              " 14  RandomForestEntr_BAG_L1_FULL   0.425000  ...       True         16\n",
              " 15           XGBoost_BAG_L1_FULL        NaN  ...       True         20\n",
              " 16      WeightedEnsemble_L2_FULL        NaN  ...       True         22\n",
              " 17   NeuralNetFastAI_BAG_L1_FULL        NaN  ...       True         12\n",
              " 18          LightGBM_BAG_L1_FULL        NaN  ...       True         14\n",
              " 19        LightGBMXT_BAG_L1_FULL        NaN  ...       True         13\n",
              " 20     LightGBMLarge_BAG_L1_FULL        NaN  ...       True         21\n",
              " 21          CatBoost_BAG_L1_FULL        NaN  ...       True         17\n",
              " \n",
              " [22 rows x 9 columns],\n",
              " 'max_stack_level': 2,\n",
              " 'model_best': 'CatBoost_BAG_L1_FULL',\n",
              " 'model_fit_times': {'CatBoost_BAG_L1': 23.49560832977295,\n",
              "  'CatBoost_BAG_L1_FULL': 0.827845573425293,\n",
              "  'ExtraTreesEntr_BAG_L1': 0.9008753299713135,\n",
              "  'ExtraTreesEntr_BAG_L1_FULL': 1.004183292388916,\n",
              "  'ExtraTreesGini_BAG_L1': 0.8981671333312988,\n",
              "  'ExtraTreesGini_BAG_L1_FULL': 0.992579460144043,\n",
              "  'LightGBMLarge_BAG_L1': 5.933805227279663,\n",
              "  'LightGBMLarge_BAG_L1_FULL': 0.23288989067077637,\n",
              "  'LightGBMXT_BAG_L1': 1.4326725006103516,\n",
              "  'LightGBMXT_BAG_L1_FULL': 0.20411396026611328,\n",
              "  'LightGBM_BAG_L1': 1.6846604347229004,\n",
              "  'LightGBM_BAG_L1_FULL': 0.1837329864501953,\n",
              "  'NeuralNetFastAI_BAG_L1': 8.605374336242676,\n",
              "  'NeuralNetFastAI_BAG_L1_FULL': 0.8907186985015869,\n",
              "  'RandomForestEntr_BAG_L1': 1.0894153118133545,\n",
              "  'RandomForestEntr_BAG_L1_FULL': 1.1021442413330078,\n",
              "  'RandomForestGini_BAG_L1': 0.9888327121734619,\n",
              "  'RandomForestGini_BAG_L1_FULL': 0.8909575939178467,\n",
              "  'WeightedEnsemble_L2': 0.2939448356628418,\n",
              "  'WeightedEnsemble_L2_FULL': 0.00031757354736328125,\n",
              "  'XGBoost_BAG_L1': 4.371852159500122,\n",
              "  'XGBoost_BAG_L1_FULL': 0.5491466522216797},\n",
              " 'model_hyperparams': {'CatBoost_BAG_L1': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': False,\n",
              "   'use_orig_features': True},\n",
              "  'CatBoost_BAG_L1_FULL': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': True,\n",
              "   'use_orig_features': True},\n",
              "  'ExtraTreesEntr_BAG_L1': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': False,\n",
              "   'use_child_oof': True,\n",
              "   'use_orig_features': True},\n",
              "  'ExtraTreesEntr_BAG_L1_FULL': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': True,\n",
              "   'use_child_oof': True,\n",
              "   'use_orig_features': True},\n",
              "  'ExtraTreesGini_BAG_L1': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': False,\n",
              "   'use_child_oof': True,\n",
              "   'use_orig_features': True},\n",
              "  'ExtraTreesGini_BAG_L1_FULL': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': True,\n",
              "   'use_child_oof': True,\n",
              "   'use_orig_features': True},\n",
              "  'LightGBMLarge_BAG_L1': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': False,\n",
              "   'use_orig_features': True},\n",
              "  'LightGBMLarge_BAG_L1_FULL': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': True,\n",
              "   'use_orig_features': True},\n",
              "  'LightGBMXT_BAG_L1': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': False,\n",
              "   'use_orig_features': True},\n",
              "  'LightGBMXT_BAG_L1_FULL': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': True,\n",
              "   'use_orig_features': True},\n",
              "  'LightGBM_BAG_L1': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': False,\n",
              "   'use_orig_features': True},\n",
              "  'LightGBM_BAG_L1_FULL': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': True,\n",
              "   'use_orig_features': True},\n",
              "  'NeuralNetFastAI_BAG_L1': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': False,\n",
              "   'use_orig_features': True},\n",
              "  'NeuralNetFastAI_BAG_L1_FULL': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': True,\n",
              "   'use_orig_features': True},\n",
              "  'RandomForestEntr_BAG_L1': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': False,\n",
              "   'use_child_oof': True,\n",
              "   'use_orig_features': True},\n",
              "  'RandomForestEntr_BAG_L1_FULL': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': True,\n",
              "   'use_child_oof': True,\n",
              "   'use_orig_features': True},\n",
              "  'RandomForestGini_BAG_L1': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': False,\n",
              "   'use_child_oof': True,\n",
              "   'use_orig_features': True},\n",
              "  'RandomForestGini_BAG_L1_FULL': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': True,\n",
              "   'use_child_oof': True,\n",
              "   'use_orig_features': True},\n",
              "  'WeightedEnsemble_L2': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': False,\n",
              "   'use_orig_features': False},\n",
              "  'WeightedEnsemble_L2_FULL': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': True,\n",
              "   'use_orig_features': False},\n",
              "  'XGBoost_BAG_L1': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': False,\n",
              "   'use_orig_features': True},\n",
              "  'XGBoost_BAG_L1_FULL': {'max_base_models': 25,\n",
              "   'max_base_models_per_type': 5,\n",
              "   'save_bag_folds': True,\n",
              "   'use_orig_features': True}},\n",
              " 'model_paths': {'CatBoost_BAG_L1': 'good_quality_cluster/models/CatBoost_BAG_L1/',\n",
              "  'CatBoost_BAG_L1_FULL': 'good_quality_cluster/models/CatBoost_BAG_L1_FULL/',\n",
              "  'ExtraTreesEntr_BAG_L1': 'good_quality_cluster/models/ExtraTreesEntr_BAG_L1/',\n",
              "  'ExtraTreesEntr_BAG_L1_FULL': 'good_quality_cluster/models/ExtraTreesEntr_BAG_L1_FULL/',\n",
              "  'ExtraTreesGini_BAG_L1': 'good_quality_cluster/models/ExtraTreesGini_BAG_L1/',\n",
              "  'ExtraTreesGini_BAG_L1_FULL': 'good_quality_cluster/models/ExtraTreesGini_BAG_L1_FULL/',\n",
              "  'LightGBMLarge_BAG_L1': 'good_quality_cluster/models/LightGBMLarge_BAG_L1/',\n",
              "  'LightGBMLarge_BAG_L1_FULL': 'good_quality_cluster/models/LightGBMLarge_BAG_L1_FULL/',\n",
              "  'LightGBMXT_BAG_L1': 'good_quality_cluster/models/LightGBMXT_BAG_L1/',\n",
              "  'LightGBMXT_BAG_L1_FULL': 'good_quality_cluster/models/LightGBMXT_BAG_L1_FULL/',\n",
              "  'LightGBM_BAG_L1': 'good_quality_cluster/models/LightGBM_BAG_L1/',\n",
              "  'LightGBM_BAG_L1_FULL': 'good_quality_cluster/models/LightGBM_BAG_L1_FULL/',\n",
              "  'NeuralNetFastAI_BAG_L1': 'good_quality_cluster/models/NeuralNetFastAI_BAG_L1/',\n",
              "  'NeuralNetFastAI_BAG_L1_FULL': 'good_quality_cluster/models/NeuralNetFastAI_BAG_L1_FULL/',\n",
              "  'RandomForestEntr_BAG_L1': 'good_quality_cluster/models/RandomForestEntr_BAG_L1/',\n",
              "  'RandomForestEntr_BAG_L1_FULL': 'good_quality_cluster/models/RandomForestEntr_BAG_L1_FULL/',\n",
              "  'RandomForestGini_BAG_L1': 'good_quality_cluster/models/RandomForestGini_BAG_L1/',\n",
              "  'RandomForestGini_BAG_L1_FULL': 'good_quality_cluster/models/RandomForestGini_BAG_L1_FULL/',\n",
              "  'WeightedEnsemble_L2': 'good_quality_cluster/models/WeightedEnsemble_L2/',\n",
              "  'WeightedEnsemble_L2_FULL': 'good_quality_cluster/models/WeightedEnsemble_L2_FULL/',\n",
              "  'XGBoost_BAG_L1': 'good_quality_cluster/models/XGBoost_BAG_L1/',\n",
              "  'XGBoost_BAG_L1_FULL': 'good_quality_cluster/models/XGBoost_BAG_L1_FULL/'},\n",
              " 'model_performance': {'CatBoost_BAG_L1': 0.6,\n",
              "  'CatBoost_BAG_L1_FULL': None,\n",
              "  'ExtraTreesEntr_BAG_L1': 0.5083333333333333,\n",
              "  'ExtraTreesEntr_BAG_L1_FULL': 0.5083333333333333,\n",
              "  'ExtraTreesGini_BAG_L1': 0.4583333333333333,\n",
              "  'ExtraTreesGini_BAG_L1_FULL': 0.4583333333333333,\n",
              "  'LightGBMLarge_BAG_L1': 0.5416666666666666,\n",
              "  'LightGBMLarge_BAG_L1_FULL': None,\n",
              "  'LightGBMXT_BAG_L1': 0.5583333333333333,\n",
              "  'LightGBMXT_BAG_L1_FULL': None,\n",
              "  'LightGBM_BAG_L1': 0.5333333333333333,\n",
              "  'LightGBM_BAG_L1_FULL': None,\n",
              "  'NeuralNetFastAI_BAG_L1': 0.5,\n",
              "  'NeuralNetFastAI_BAG_L1_FULL': None,\n",
              "  'RandomForestEntr_BAG_L1': 0.425,\n",
              "  'RandomForestEntr_BAG_L1_FULL': 0.425,\n",
              "  'RandomForestGini_BAG_L1': 0.4666666666666667,\n",
              "  'RandomForestGini_BAG_L1_FULL': 0.4666666666666667,\n",
              "  'WeightedEnsemble_L2': 0.6,\n",
              "  'WeightedEnsemble_L2_FULL': None,\n",
              "  'XGBoost_BAG_L1': 0.525,\n",
              "  'XGBoost_BAG_L1_FULL': None},\n",
              " 'model_pred_times': {'CatBoost_BAG_L1': 0.03770184516906738,\n",
              "  'CatBoost_BAG_L1_FULL': None,\n",
              "  'ExtraTreesEntr_BAG_L1': 0.1131281852722168,\n",
              "  'ExtraTreesEntr_BAG_L1_FULL': 0.11022233963012695,\n",
              "  'ExtraTreesGini_BAG_L1': 0.10319185256958008,\n",
              "  'ExtraTreesGini_BAG_L1_FULL': 0.11887097358703613,\n",
              "  'LightGBMLarge_BAG_L1': 0.025969505310058594,\n",
              "  'LightGBMLarge_BAG_L1_FULL': None,\n",
              "  'LightGBMXT_BAG_L1': 0.023320913314819336,\n",
              "  'LightGBMXT_BAG_L1_FULL': None,\n",
              "  'LightGBM_BAG_L1': 0.0206758975982666,\n",
              "  'LightGBM_BAG_L1_FULL': None,\n",
              "  'NeuralNetFastAI_BAG_L1': 0.19253849983215332,\n",
              "  'NeuralNetFastAI_BAG_L1_FULL': None,\n",
              "  'RandomForestEntr_BAG_L1': 0.1051628589630127,\n",
              "  'RandomForestEntr_BAG_L1_FULL': 0.12531399726867676,\n",
              "  'RandomForestGini_BAG_L1': 0.10991787910461426,\n",
              "  'RandomForestGini_BAG_L1_FULL': 0.1143655776977539,\n",
              "  'WeightedEnsemble_L2': 0.0005311965942382812,\n",
              "  'WeightedEnsemble_L2_FULL': 0.0006184577941894531,\n",
              "  'XGBoost_BAG_L1': 0.03607344627380371,\n",
              "  'XGBoost_BAG_L1_FULL': None},\n",
              " 'model_types': {'CatBoost_BAG_L1': 'StackerEnsembleModel_CatBoost',\n",
              "  'CatBoost_BAG_L1_FULL': 'StackerEnsembleModel_CatBoost',\n",
              "  'ExtraTreesEntr_BAG_L1': 'StackerEnsembleModel_XT',\n",
              "  'ExtraTreesEntr_BAG_L1_FULL': 'StackerEnsembleModel_XT',\n",
              "  'ExtraTreesGini_BAG_L1': 'StackerEnsembleModel_XT',\n",
              "  'ExtraTreesGini_BAG_L1_FULL': 'StackerEnsembleModel_XT',\n",
              "  'LightGBMLarge_BAG_L1': 'StackerEnsembleModel_LGB',\n",
              "  'LightGBMLarge_BAG_L1_FULL': 'StackerEnsembleModel_LGB',\n",
              "  'LightGBMXT_BAG_L1': 'StackerEnsembleModel_LGB',\n",
              "  'LightGBMXT_BAG_L1_FULL': 'StackerEnsembleModel_LGB',\n",
              "  'LightGBM_BAG_L1': 'StackerEnsembleModel_LGB',\n",
              "  'LightGBM_BAG_L1_FULL': 'StackerEnsembleModel_LGB',\n",
              "  'NeuralNetFastAI_BAG_L1': 'StackerEnsembleModel_NNFastAiTabular',\n",
              "  'NeuralNetFastAI_BAG_L1_FULL': 'StackerEnsembleModel_NNFastAiTabular',\n",
              "  'RandomForestEntr_BAG_L1': 'StackerEnsembleModel_RF',\n",
              "  'RandomForestEntr_BAG_L1_FULL': 'StackerEnsembleModel_RF',\n",
              "  'RandomForestGini_BAG_L1': 'StackerEnsembleModel_RF',\n",
              "  'RandomForestGini_BAG_L1_FULL': 'StackerEnsembleModel_RF',\n",
              "  'WeightedEnsemble_L2': 'WeightedEnsembleModel',\n",
              "  'WeightedEnsemble_L2_FULL': 'WeightedEnsembleModel',\n",
              "  'XGBoost_BAG_L1': 'StackerEnsembleModel_XGBoost',\n",
              "  'XGBoost_BAG_L1_FULL': 'StackerEnsembleModel_XGBoost'},\n",
              " 'num_bag_folds': 5,\n",
              " 'num_classes': 3}"
            ]
          },
          "metadata": {},
          "execution_count": 136
        }
      ],
      "source": [
        "predictor.fit_summary()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "JpD-GF51_0rr"
      },
      "outputs": [],
      "source": [
        "test_data = x_test.loc[:, selected_genes].copy()\n",
        "y_pred_score = predictor.predict_proba(test_data)\n",
        "y_pred = predictor.predict(test_data)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "HdVoYC5TDIHC"
      },
      "outputs": [],
      "source": [
        "from sklearn.metrics import (accuracy_score, balanced_accuracy_score, \n",
        "                             roc_auc_score, f1_score)\n",
        "def performance_scores(y_true, y_pred_score, y_pred=None):\n",
        "    # We can find which class has the highest score as its predicted class\n",
        "    if y_pred is None:\n",
        "        y_pred = y_pred_score.argmax(axis=1)\n",
        "        \n",
        "    return {\n",
        "        \"accuracy\": accuracy_score(y_true, y_pred),\n",
        "        \"balanced_accuracy\": balanced_accuracy_score(y_true, y_pred),\n",
        "        \"auroc\": roc_auc_score(y_true, y_pred_score, average=\"weighted\",\n",
        "                               multi_class=\"ovr\"),\n",
        "        \"f1\": f1_score(y_true, y_pred, average=\"weighted\")\n",
        "    }"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hO597nTKDDAJ",
        "outputId": "5bd26423-615d-413d-a411-48d6ab7d49a1"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'accuracy': 0.5,\n",
              " 'auroc': 0.5679951690821257,\n",
              " 'balanced_accuracy': 0.38095238095238093,\n",
              " 'f1': 0.36296296296296293}"
            ]
          },
          "metadata": {},
          "execution_count": 143
        }
      ],
      "source": [
        "performance_scores(y_test, y_pred_score, y_pred)"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "name": "Project_v2.ipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}